{
 "metadata": {
  "name": "",
  "signature": "sha256:958480fba8b0e7f3a46b5660b954fcb85a44c6839181a62e45f4a53e1ef30574"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "#Prepare data from snOPY database"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "##Goal\n",
      "\n",
      "In order to unify our calculations of snoRNA target sites we decided to change database from [snoRNABase](https://www-snorna.biotoul.fr/) to [snOPY](http://snoopy.med.miyazaki-u.ac.jp/snorna_db.cgi). This is because snoRNABase does not have any annotation of mouse snoRNAs and it is a bit outdated. We kindly obtained xml files with database entries from dr Kenmochi from Frontier Science Research Center, University of Miyazaki, Japan. With this in hand we would be able to build an uniform pipeline for both species."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "##Materials and Methods"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "###Read and wrangle the data"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "An example of XML input file is shown below:"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "<snoopy id='Homo_sapiens300000'>\n",
      "<dataquality>B</dataquality>\n",
      "<openstatus>+</openstatus>\n",
      "<lastupdate>2009-07-01</lastupdate>\n",
      "<blindnote>\n",
      "gene name:AC013603.17->SNORA70B-pseud.\n",
      "\n",
      "2009\u5e747\u67081\u65e5\uff08\u5c71\u5143\uff09gene name\uff1aSNORA70B-pseud\u2192SNORA70BL3\n",
      "\n",
      "\n",
      "</blindnote>\n",
      "<organism>Homo_sapiens</organism>\n",
      "<gene_name>SNORA70BL3</gene_name>\n",
      "<family_name></family_name>\n",
      "<alias></alias>\n",
      "<family_base>Homo_sapiens</family_base>\n",
      "<chrom_or_contig>8</chrom_or_contig>\n",
      "<genome_position>33517083..33517197</genome_position>\n",
      "<strand>+</strand>\n",
      "<box type='c'></box>\n",
      "<box type='d'></box>\n",
      "<box type='h'>4..9</box>\n",
      "<box type='aca'>110..112</box>\n",
      "<host_id>Homo_sapiens100000</host_id>\n",
      "<host_gene>Mono:286:AC013603.17</host_gene>\n",
      "<host_intron></host_intron>\n",
      "<host_position>1001..1115</host_position>\n",
      "<organization>Mono</organization>\n",
      "<mod_type>H/ACA</mod_type>\n",
      "<mod_rna></mod_rna>\n",
      "<mod_site></mod_site>\n",
      "<duplex_region></duplex_region>\n",
      "<accession></accession>\n",
      "<evidence></evidence>\n",
      "<note></note>\n",
      "<sequence>\n",
      "TTGATAAGAAAGAAGTTCTGATCCCAGTGTGCAATGGCTGCAAACAGCAGCTTTCTTGGT\n",
      "GGTGTACATGGCCTGTTTCTTGTATGGGTTGCTCTAAGGGTCCTTGGAGACAGGC\n",
      "</sequence>\n",
      "</snoopy>"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data = {}\n",
      "for xml in glob.glob(\"./snOPY/snoRNAs/*.xml\"):\n",
      "    try:\n",
      "        tree = ET.parse(xml)\n",
      "        root = tree.getroot()\n",
      "    except ET.ParseError, e:\n",
      "        print \"Cannot parse: \", xml, e\n",
      "    tmp = {}\n",
      "    for child in root:\n",
      "        if child.tag == \"box\":\n",
      "            if child.text:\n",
      "                tmp[child.tag + \"_\" + child.attrib['type']] = child.text\n",
      "            else:\n",
      "                tmp[child.tag] = np.nan\n",
      "        else:\n",
      "            if child.text:\n",
      "                tmp[child.tag] = child.text.replace(\"\\n\", \"\")\n",
      "            else:\n",
      "                tmp[child.tag] = np.nan\n",
      "    data[root.attrib[\"id\"]] = tmp"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 206
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df = pd.DataFrame(data).T\n",
      "df[\"id\"] = [i.split(\"_\")[-1] for i in df.index.tolist()]\n",
      "df = df.dropna(subset=[\"mod_type\", \"genome_position\"])\n",
      "df[\"start\"] = [int(i.split(\"..\")[0]) - 1 for i in df[\"genome_position\"]]\n",
      "df[\"end\"] = [int(i.split(\"..\")[1]) for i in df[\"genome_position\"]]\n",
      "df[\"snor_id\"] = df.gene_name + \"_\" + df.id\n",
      "df = df[[\"chrom_or_contig\",\n",
      "    \"start\",\n",
      "    \"end\",\n",
      "    \"snor_id\",\n",
      "    \"mod_type\",\n",
      "    \"strand\",\n",
      "    \"sequence\",\n",
      "    \"box_d\",\n",
      "    \"box_c\",\n",
      "    \"box_h\",\n",
      "    \"box_aca\",\n",
      "    \"alias\",\n",
      "    \"gene_name\",\n",
      "    \"accession\",\n",
      "    \"mod_site\",\n",
      "    \"host_gene\",\n",
      "    \"host_id\",\n",
      "    \"organization\",\n",
      "    \"organism\",\n",
      "    \"note\"]]\n",
      "df.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>chrom_or_contig</th>\n",
        "      <th>start</th>\n",
        "      <th>end</th>\n",
        "      <th>snor_id</th>\n",
        "      <th>mod_type</th>\n",
        "      <th>strand</th>\n",
        "      <th>sequence</th>\n",
        "      <th>box_d</th>\n",
        "      <th>box_c</th>\n",
        "      <th>box_h</th>\n",
        "      <th>box_aca</th>\n",
        "      <th>alias</th>\n",
        "      <th>gene_name</th>\n",
        "      <th>accession</th>\n",
        "      <th>mod_site</th>\n",
        "      <th>host_gene</th>\n",
        "      <th>host_id</th>\n",
        "      <th>organization</th>\n",
        "      <th>organism</th>\n",
        "      <th>note</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>Homo_sapiens300000</th>\n",
        "      <td>  8</td>\n",
        "      <td>  33517082</td>\n",
        "      <td>  33517197</td>\n",
        "      <td> SNORA70BL3_sapiens300000</td>\n",
        "      <td> H/ACA</td>\n",
        "      <td> +</td>\n",
        "      <td> TTGATAAGAAAGAAGTTCTGATCCCAGTGTGCAATGGCTGCAAACA...</td>\n",
        "      <td> NaN</td>\n",
        "      <td> NaN</td>\n",
        "      <td>   4..9</td>\n",
        "      <td> 110..112</td>\n",
        "      <td> NaN</td>\n",
        "      <td> SNORA70BL3</td>\n",
        "      <td> NaN</td>\n",
        "      <td> NaN</td>\n",
        "      <td> Mono:286:AC013603.17</td>\n",
        "      <td> Homo_sapiens100000</td>\n",
        "      <td> Mono</td>\n",
        "      <td> Homo_sapiens</td>\n",
        "      <td> NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Homo_sapiens300001</th>\n",
        "      <td> 12</td>\n",
        "      <td>  55541425</td>\n",
        "      <td>  55541559</td>\n",
        "      <td>  SNORA48L6_sapiens300001</td>\n",
        "      <td> H/ACA</td>\n",
        "      <td> +</td>\n",
        "      <td> TGTTCCCGATCTGGTAGAGTAGCATCTGGTTGGTGGTGACCATCTA...</td>\n",
        "      <td> NaN</td>\n",
        "      <td> NaN</td>\n",
        "      <td> 16..21</td>\n",
        "      <td> 129..131</td>\n",
        "      <td> NaN</td>\n",
        "      <td>  SNORA48L6</td>\n",
        "      <td> NaN</td>\n",
        "      <td> NaN</td>\n",
        "      <td> Mono:245:AC121758.14</td>\n",
        "      <td> Homo_sapiens100001</td>\n",
        "      <td> Mono</td>\n",
        "      <td> Homo_sapiens</td>\n",
        "      <td> NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Homo_sapiens300002</th>\n",
        "      <td>  2</td>\n",
        "      <td> 139985463</td>\n",
        "      <td> 139985595</td>\n",
        "      <td>  SNORA72L7_sapiens300002</td>\n",
        "      <td> H/ACA</td>\n",
        "      <td> +</td>\n",
        "      <td> CTGTGAATATTCTCTGTGTTCTGATTATGTGAGAGCCAGGACAGGC...</td>\n",
        "      <td> NaN</td>\n",
        "      <td> NaN</td>\n",
        "      <td> 59..64</td>\n",
        "      <td>   59..61</td>\n",
        "      <td> NaN</td>\n",
        "      <td>  SNORA72L7</td>\n",
        "      <td> NaN</td>\n",
        "      <td> NaN</td>\n",
        "      <td>   Mono:45:AC016710.7</td>\n",
        "      <td> Homo_sapiens100002</td>\n",
        "      <td> Mono</td>\n",
        "      <td> Homo_sapiens</td>\n",
        "      <td> NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Homo_sapiens300003</th>\n",
        "      <td>  7</td>\n",
        "      <td>  12706907</td>\n",
        "      <td>  12707039</td>\n",
        "      <td>  SNORA64L2_sapiens300003</td>\n",
        "      <td> H/ACA</td>\n",
        "      <td> -</td>\n",
        "      <td> ACACTCTCAGCTATGTATAGTTACACCTGGCTTCACTTGTGTGACT...</td>\n",
        "      <td> NaN</td>\n",
        "      <td> NaN</td>\n",
        "      <td> 65..70</td>\n",
        "      <td> 127..129</td>\n",
        "      <td> NaN</td>\n",
        "      <td>  SNORA64L2</td>\n",
        "      <td> NaN</td>\n",
        "      <td> NaN</td>\n",
        "      <td>  Mono:182:AC011891.3</td>\n",
        "      <td> Homo_sapiens100003</td>\n",
        "      <td> Mono</td>\n",
        "      <td> Homo_sapiens</td>\n",
        "      <td> NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Homo_sapiens300004</th>\n",
        "      <td> 19</td>\n",
        "      <td>   9791642</td>\n",
        "      <td>   9791770</td>\n",
        "      <td>  SNORA70L2_sapiens300004</td>\n",
        "      <td> H/ACA</td>\n",
        "      <td> +</td>\n",
        "      <td> CTGAATTGGCAAGTCCACCAAGAAACTATAAAAATGTGTGTGTGCA...</td>\n",
        "      <td> NaN</td>\n",
        "      <td> NaN</td>\n",
        "      <td> 23..28</td>\n",
        "      <td> 123..125</td>\n",
        "      <td> NaN</td>\n",
        "      <td>  SNORA70L2</td>\n",
        "      <td> NaN</td>\n",
        "      <td> NaN</td>\n",
        "      <td>  Mono:137:AC008752.6</td>\n",
        "      <td> Homo_sapiens100004</td>\n",
        "      <td> Mono</td>\n",
        "      <td> Homo_sapiens</td>\n",
        "      <td> NaN</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>5 rows \u00d7 20 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 207,
       "text": [
        "                   chrom_or_contig      start        end  \\\n",
        "Homo_sapiens300000               8   33517082   33517197   \n",
        "Homo_sapiens300001              12   55541425   55541559   \n",
        "Homo_sapiens300002               2  139985463  139985595   \n",
        "Homo_sapiens300003               7   12706907   12707039   \n",
        "Homo_sapiens300004              19    9791642    9791770   \n",
        "\n",
        "                                     snor_id mod_type strand  \\\n",
        "Homo_sapiens300000  SNORA70BL3_sapiens300000    H/ACA      +   \n",
        "Homo_sapiens300001   SNORA48L6_sapiens300001    H/ACA      +   \n",
        "Homo_sapiens300002   SNORA72L7_sapiens300002    H/ACA      +   \n",
        "Homo_sapiens300003   SNORA64L2_sapiens300003    H/ACA      -   \n",
        "Homo_sapiens300004   SNORA70L2_sapiens300004    H/ACA      +   \n",
        "\n",
        "                                                             sequence box_d  \\\n",
        "Homo_sapiens300000  TTGATAAGAAAGAAGTTCTGATCCCAGTGTGCAATGGCTGCAAACA...   NaN   \n",
        "Homo_sapiens300001  TGTTCCCGATCTGGTAGAGTAGCATCTGGTTGGTGGTGACCATCTA...   NaN   \n",
        "Homo_sapiens300002  CTGTGAATATTCTCTGTGTTCTGATTATGTGAGAGCCAGGACAGGC...   NaN   \n",
        "Homo_sapiens300003  ACACTCTCAGCTATGTATAGTTACACCTGGCTTCACTTGTGTGACT...   NaN   \n",
        "Homo_sapiens300004  CTGAATTGGCAAGTCCACCAAGAAACTATAAAAATGTGTGTGTGCA...   NaN   \n",
        "\n",
        "                   box_c   box_h   box_aca alias   gene_name accession  \\\n",
        "Homo_sapiens300000   NaN    4..9  110..112   NaN  SNORA70BL3       NaN   \n",
        "Homo_sapiens300001   NaN  16..21  129..131   NaN   SNORA48L6       NaN   \n",
        "Homo_sapiens300002   NaN  59..64    59..61   NaN   SNORA72L7       NaN   \n",
        "Homo_sapiens300003   NaN  65..70  127..129   NaN   SNORA64L2       NaN   \n",
        "Homo_sapiens300004   NaN  23..28  123..125   NaN   SNORA70L2       NaN   \n",
        "\n",
        "                   mod_site             host_gene             host_id  \\\n",
        "Homo_sapiens300000      NaN  Mono:286:AC013603.17  Homo_sapiens100000   \n",
        "Homo_sapiens300001      NaN  Mono:245:AC121758.14  Homo_sapiens100001   \n",
        "Homo_sapiens300002      NaN    Mono:45:AC016710.7  Homo_sapiens100002   \n",
        "Homo_sapiens300003      NaN   Mono:182:AC011891.3  Homo_sapiens100003   \n",
        "Homo_sapiens300004      NaN   Mono:137:AC008752.6  Homo_sapiens100004   \n",
        "\n",
        "                   organization      organism note  \n",
        "Homo_sapiens300000         Mono  Homo_sapiens  NaN  \n",
        "Homo_sapiens300001         Mono  Homo_sapiens  NaN  \n",
        "Homo_sapiens300002         Mono  Homo_sapiens  NaN  \n",
        "Homo_sapiens300003         Mono  Homo_sapiens  NaN  \n",
        "Homo_sapiens300004         Mono  Homo_sapiens  NaN  \n",
        "\n",
        "[5 rows x 20 columns]"
       ]
      }
     ],
     "prompt_number": 207
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "###Write columns help"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "desc_map = OrderedDict(((\"Chromosome\", \"Chromosome on which gene is located\"),\n",
      "            (\"Start\", \"0-based start coordinate\"),\n",
      "            (\"End\", \"1-based end coordinate\"),\n",
      "            (\"snoRNA ID\", \"ID for snoRNA\"),\n",
      "            (\"Type\", \"Type of snoRNA (C/D or H/ACA)\"),\n",
      "            (\"Strand\", \"Strand on which gene is located\"),\n",
      "            (\"Sequence\", \"Sequence of snoRNA\"),\n",
      "            (\"D-boxes\", \"1-based coordinates snoRNA D-boxes in form of \"+\n",
      "                       \"D-box_start..D-box_end or D'-box_start..D'-box_end,D-box_start..D-box_end \"+\n",
      "                       \"eg. 60..63 or 30..33,60..63. It must be 4 nucleotides long.\"),\n",
      "            (\"C-boxes\", \"Same as above (not that prime boxes are reversed in positions) \"+\n",
      "                       \"with exception of the length constraint\"),\n",
      "            (\"H-box\", \"1-based position of the H box in H/ACA box snoRNAs in form of start..end\"),\n",
      "            (\"ACA-box\", \"1-based position of the ACA box in H/ACA box snoRNA in form of start..end\"),\n",
      "            (\"Alias\", \"Alias/Alternative name for snoRNA\"),\n",
      "            (\"Gene name\", \"Name of the snoRNA gene\"),\n",
      "            (\"Accession\", \"Accession (NCBI, RefSeq etc) for snoRNA\"),\n",
      "            (\"Modification sites\", \"Coma separated list of modified sites in form of \"+\n",
      "                                  \"ModifiedRNA:NucleotidePosition eg. 28S:U46,18S:G52\"),\n",
      "            (\"Host gene\", \"Host gene name if any\"),\n",
      "            (\"Host ID\", \"Host locus ID\"),\n",
      "            (\"Organization\", \"Organisation of the gene eg. intronic\"),\n",
      "            (\"Organism\", \"Species\"),\n",
      "            (\"Note\", \"Additional info about snoRNA\"))\n",
      ")\n",
      "with open(\"Columns_description.tab\", \"w\") as col_desc:\n",
      "    counter = 1\n",
      "    for col, desc in desc_map.iteritems():\n",
      "        col_desc.write(\"%i. %s\\t%s\\n\" % (counter,\n",
      "                                         col,\n",
      "                                         desc))\n",
      "        counter += 1"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 182
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "###Homo sapiens"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "homo_sapiens_cd = df[(df.organism == \"Homo_sapiens\") & (df.mod_type==\"C/D\")]\n",
      "homo_sapiens_cd = homo_sapiens_cd.dropna(subset=[\"box_d\"])\n",
      "homo_sapiens_aca = df[(df.organism == \"Homo_sapiens\") & (df.mod_type==\"H/ACA\")]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 162
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "homo_sapiens_cd.to_csv(\"human/hsa_cd_box_snoRNAs.tab\",\n",
      "                       index=False,\n",
      "                       header=False,\n",
      "                       sep=\"\\t\",\n",
      "                       na_rep=\"NaN\")\n",
      "homo_sapiens_aca.to_csv(\"human/hsa_haca_box_snoRNAs.tab\",\n",
      "                       index=False,\n",
      "                       header=False,\n",
      "                       sep=\"\\t\",\n",
      "                       na_rep=\"NaN\",\n",
      "                        encoding='utf-8')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 187
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "###Mus musculus"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "mus_musculus_cd = df[(df.organism == \"Mus_musculus\") & (df.mod_type==\"C/D\")]\n",
      "mus_musculus_cd = mus_musculus_cd.dropna(subset=[\"box_d\"])\n",
      "mus_musculus_aca = df[(df.organism == \"Mus_musculus\") & (df.mod_type==\"H/ACA\")]\n",
      "cd_for_conversion = df[(df.organism == \"Mus_musculus\") & (df.mod_type==\"C/D\")]\n",
      "cd_for_conversion[\"chrom_or_contig\"] = \"chr\" + cd_for_conversion.chrom_or_contig\n",
      "cd_for_conversion.iloc[:,:6].to_csv(\"mouse/mmu_mm9_cd_box_snoRNAs.bed\",\n",
      "                                    header=None,\n",
      "                                    sep=\"\\t\",\n",
      "                                    index=None)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 31
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "####Convert mm9 coordinates to mm10"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash\n",
      "CrossMap.py bed ~/Data/Genome/genome_conversion_files/mm9ToMm10.over.chain.gz \\\n",
      "                mouse/mmu_mm9_cd_box_snoRNAs.bed | \\\n",
      "                cut -f 8,9,10,11,12,13,14 > \\\n",
      "                mouse/mmu_mm10_cd_box_snoRNAs.bed"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "@ 2015-02-27 16:16:05: Read chain_file:  /import/bc2/home/zavolan/gumiennr/Data/Genome/genome_conversion_files/mm9ToMm10.over.chain.gz\n",
        "End corrdinate is not an integer. skip 166888972\t166889065\tSNORD12_musculus300988\tC/D\t+\n"
       ]
      }
     ],
     "prompt_number": 21
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "mm9_2_mm10 = pd.read_table(\"mouse/mmu_mm10_cd_box_snoRNAs.bed\", index_col=3,\n",
      "                           header=None)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 24
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "mm9_2_mm10.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>2</th>\n",
        "      <th>4</th>\n",
        "      <th>5</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>SNORD86_musculus300006</th>\n",
        "      <td> chr14</td>\n",
        "      <td>  56837070</td>\n",
        "      <td>  56837156</td>\n",
        "      <td> C/D</td>\n",
        "      <td> -</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>SNORD23_musculus300008</th>\n",
        "      <td> chr10</td>\n",
        "      <td>  21878783</td>\n",
        "      <td>  21878892</td>\n",
        "      <td> C/D</td>\n",
        "      <td> -</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>SNORND104_musculus300032</th>\n",
        "      <td> chr11</td>\n",
        "      <td> 106500992</td>\n",
        "      <td> 106501062</td>\n",
        "      <td> C/D</td>\n",
        "      <td> +</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>SNORD116_musculus300055</th>\n",
        "      <td>  chr7</td>\n",
        "      <td>  59859213</td>\n",
        "      <td>  59859307</td>\n",
        "      <td> C/D</td>\n",
        "      <td> -</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>SNORD116_musculus300056</th>\n",
        "      <td>  chr7</td>\n",
        "      <td>  59861730</td>\n",
        "      <td>  59861824</td>\n",
        "      <td> C/D</td>\n",
        "      <td> -</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>5 rows \u00d7 5 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 27,
       "text": [
        "                              0          1          2    4  5\n",
        "3                                                            \n",
        "SNORD86_musculus300006    chr14   56837070   56837156  C/D  -\n",
        "SNORD23_musculus300008    chr10   21878783   21878892  C/D  -\n",
        "SNORND104_musculus300032  chr11  106500992  106501062  C/D  +\n",
        "SNORD116_musculus300055    chr7   59859213   59859307  C/D  -\n",
        "SNORD116_musculus300056    chr7   59861730   59861824  C/D  -\n",
        "\n",
        "[5 rows x 5 columns]"
       ]
      }
     ],
     "prompt_number": 27
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "mus_musculus_cd[\"start\"] = [int(mm9_2_mm10.ix[i, 1]) if i in mm9_2_mm10.index\n",
      "                            else np.nan for i in mus_musculus_cd.snor_id]\n",
      "mus_musculus_cd[\"end\"] = [int(mm9_2_mm10.ix[i, 2]) if i in mm9_2_mm10.index\n",
      "                            else np.nan for i in mus_musculus_cd.snor_id]\n",
      "mus_musculus_cd = mus_musculus_cd.dropna(subset=[\"start\"])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 34
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "mus_musculus_cd.to_csv(\"mouse/mmu_mm10_cd_box_snoRNAs.tab\",\n",
      "                       index=False,\n",
      "                       header=False,\n",
      "                       sep=\"\\t\",\n",
      "                       na_rep=\"NaN\")\n",
      "mus_musculus_aca.to_csv(\"mouse/mmu__mm10_haca_box_snoRNAs.tab\",\n",
      "                       index=False,\n",
      "                       header=False,\n",
      "                       sep=\"\\t\",\n",
      "                       na_rep=\"NaN\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 37
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "##Utils"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%matplotlib inline\n",
      "import re\n",
      "import glob\n",
      "import pandas as pd\n",
      "import numpy as np\n",
      "import pylab as pl\n",
      "import seaborn as sns\n",
      "from random import choice\n",
      "import xml.etree.ElementTree as ET\n",
      "from collections import OrderedDict\n",
      "from Bio import SeqIO\n",
      "from IPython.display import Image\n",
      "pl.rcParams['figure.figsize'] = (14, 10)\n",
      "pl.rcParams['ytick.labelsize'] = 20\n",
      "pl.rcParams['xtick.labelsize'] = 20\n",
      "pl.rcParams['axes.labelsize'] = 23\n",
      "pl.rcParams['legend.fontsize'] = 20\n",
      "sns.set_style('ticks')\n",
      "c1, c2, c3 = sns.color_palette(\"Set1\", 3)\n",
      "from IPython.core.display import HTML, Image\n",
      "css_file = '/import/bc2/home/zavolan/gumiennr/.ipython/profile_default/static/custom/custom.css'\n",
      "HTML(open(css_file, \"r\").read())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "/*\n",
        " * example.css\n",
        " * Copyright (C) 2014 Rafal Gumienny\n",
        " *\n",
        " * Distributed under terms of the GPL license.\n",
        " */\n",
        "\n",
        "<link href='http://fonts.googleapis.com/css?family=Alegreya+Sans:100,300,400,500,700,800,900,100italic,300italic,400italic,500italic,700italic,800italic,900italic' rel='stylesheet' type='text/css'>\n",
        "<link href='http://fonts.googleapis.com/css?family=Arvo:400,700,400italic' rel='stylesheet' type='text/css'>\n",
        "<link href='http://fonts.googleapis.com/css?family=PT+Mono' rel='stylesheet' type='text/css'>\n",
        "<link href='http://fonts.googleapis.com/css?family=Shadows+Into+Light' rel='stylesheet' type='text/css'>\n",
        "<link href='http://fonts.googleapis.com/css?family=Philosopher:400,700,400italic,700italic' rel='stylesheet' type='text/css'>\n",
        "\n",
        "<style>\n",
        "\n",
        "@font-face {\n",
        "\tfont-family: \"Computer Modern\";\n",
        "\tsrc: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');\n",
        "}\n",
        "\n",
        "div.cell {\n",
        "\twidth: 800px;\n",
        "\tmargin-left: auto;\n",
        "\tmargin-right: auto;\n",
        "}\n",
        "\n",
        "#notebook li {\n",
        "margin-top:0.6em;\n",
        "}\n",
        "\n",
        "\n",
        "/* [> draw border around running cells <] */\n",
        "div.cell.border-box-sizing.code_cell.running { \n",
        "\tborder: 3px solid #111;\n",
        "\tborder-color: #B22222;\n",
        "\tborder-radius: 5px;\n",
        "}\n",
        "\n",
        "div.text_cell_render{\n",
        "    font-family: Computer Modern, \"Helvetica Neue\", Arial, Helvetica, Geneva, sans-serif;\n",
        "    line-height: 145%;\n",
        "    font-size: 130%;\n",
        "    /* margin-left:10px; */\n",
        "    /* margin-right:10px; */\n",
        "}\n",
        "\n",
        "/* Formatting for header cells */\n",
        ".text_cell_render h1 {\n",
        "\tfont-family: 'Serif', sans-serif;\n",
        "\tfont-weight: 400;\n",
        "\tfont-size: 30pt;\n",
        "\tline-height: 100%;\n",
        "\t/* color: rgb(12,85,97); */\n",
        "\tcolor: #b22222;\n",
        "\tmargin-bottom: 0.1em;\n",
        "\tmargin-top: 0.1em;\n",
        "\tdisplay: block;\n",
        "}\t\n",
        ".text_cell_render h2 {\n",
        "\tfont-family: 'Serif', serif;\n",
        "\tfont-weight: 700;\n",
        "\tfont-size: 20pt;\n",
        "\tline-height: 100%;\n",
        "\tcolor: #00008B;\n",
        "\t/* color: rgb(12,85,97); */\n",
        "\t/* color: rgb(171,165,131); */\n",
        "\tmargin-bottom: 0.1em;\n",
        "\tmargin-top: 0.3em;\n",
        "\tdisplay: block;\n",
        "}\t\n",
        "\n",
        ".text_cell_render h3 {\n",
        "\tfont-family: 'Serif', serif;\n",
        "\tmargin-bottom: 0.1em;\n",
        "\tmargin-top: 0.3em;\n",
        "\t/* margin-top:12px; */\n",
        "\t/* margin-bottom: 1px; */\n",
        "\tfont-size: 15pt;\n",
        "\tfont-style: italic;\n",
        "\tcolor: #004b00;\n",
        "\t/* color: rgb(95,92,72); */\n",
        "}\n",
        "\n",
        ".text_cell_render h4 {\n",
        "\tfont-family: serif;\n",
        "\tmargin-bottom: 0.1em;\n",
        "\tmargin-top: 0.3em;\n",
        "\t/* margin-top:12px; */\n",
        "\t/* margin-bottom: 1px; */\n",
        "}\n",
        "\n",
        ".text_cell_render h5 {\n",
        "\tfont-family: sans-serif;\n",
        "\tfont-weight: 300;\n",
        "\tfont-size: 13pt;\n",
        "\tcolor: grey;\n",
        "\tfont-style: italic;\n",
        "\tmargin-bottom: .1em;\n",
        "\tmargin-top: 0.3em;\n",
        "\tdisplay: block;\n",
        "}\n",
        "\n",
        ".text_cell_render h6 {\n",
        "\tfont-family: sans-serif;\n",
        "\tfont-weight: 300;\n",
        "\tfont-size: 10pt;\n",
        "\tcolor: grey;\n",
        "\tmargin-bottom: 1px;\n",
        "\tmargin-top: 1px;\n",
        "}\n",
        "\n",
        ".CodeMirror{\n",
        "        font-family: 'DejaVu Sans Mono';\n",
        "        font-size: 100%;\n",
        "}\n",
        "\n",
        "table\n",
        "{\n",
        "    font-family: \"Lucida Sans Unicode\", \"Lucida Grande\", Sans-Serif;\n",
        "    font-size: 14px;\n",
        "    margin: 45px;\n",
        "    text-align: center;\n",
        "    border-collapse: collapse;\n",
        "}\n",
        "table th\n",
        "{\n",
        "    font-size: 15px;\n",
        "    font-weight: normal;\n",
        "    padding: 8px;\n",
        "    background: #DCDCDC;\n",
        "    border-top: 4px solid #aabcfe;\n",
        "    border-bottom: 1px solid #fff;\n",
        "    color: #000000;\n",
        "\ttext-align: center;\n",
        "}\n",
        "table td\n",
        "{\n",
        "    padding: 8px;\n",
        "    background: #FFFACD;\n",
        "    border-bottom: 1px solid #fff;\n",
        "    color: #000000;\n",
        "    border-top: 1px solid transparent;\n",
        "}\n",
        "table tr:hover td\n",
        "{\n",
        "    background: #d0dafd;\n",
        "    color: #000000;\n",
        "}\n",
        "\n",
        "</style>\n",
        "\n"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 195,
       "text": [
        "<IPython.core.display.HTML at 0x7350210>"
       ]
      }
     ],
     "prompt_number": 195
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}