{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Prepare data from snOPY database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Goal\n",
    "\n",
    "In order to unify our calculations of snoRNA target sites we decided to change database from [snoRNABase](https://www-snorna.biotoul.fr/) to [snOPY](http://snoopy.med.miyazaki-u.ac.jp/snorna_db.cgi). This is because snoRNABase does not have any annotation of mouse snoRNAs and it is a bit outdated. We kindly obtained xml files with database entries from dr Kenmochi from Frontier Science Research Center, University of Miyazaki, Japan. With this in hand we would be able to build an uniform pipeline for both species.\n",
    "\n",
    "The biggest disadvantage of this database is that the coordinates are mixed i.e. for human coordinates consist of hg19 and hg18 assemblies so we need to identify them and lift over hg18 to hg19. The same is might be true for mouse were there is a mix of mm9 and mm10 coordinates."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Materials and Methods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Read and wrangle the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "An example of XML input file is shown below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "<snoopy id='Homo_sapiens300000'>\n",
    "<dataquality>B</dataquality>\n",
    "<openstatus>+</openstatus>\n",
    "<lastupdate>2009-07-01</lastupdate>\n",
    "<blindnote>\n",
    "gene name:AC013603.17->SNORA70B-pseud.\n",
    "\n",
    "2009年7月1日（山元）gene name：SNORA70B-pseud→SNORA70BL3\n",
    "\n",
    "\n",
    "</blindnote>\n",
    "<organism>Homo_sapiens</organism>\n",
    "<gene_name>SNORA70BL3</gene_name>\n",
    "<family_name></family_name>\n",
    "<alias></alias>\n",
    "<family_base>Homo_sapiens</family_base>\n",
    "<chrom_or_contig>8</chrom_or_contig>\n",
    "<genome_position>33517083..33517197</genome_position>\n",
    "<strand>+</strand>\n",
    "<box type='c'></box>\n",
    "<box type='d'></box>\n",
    "<box type='h'>4..9</box>\n",
    "<box type='aca'>110..112</box>\n",
    "<host_id>Homo_sapiens100000</host_id>\n",
    "<host_gene>Mono:286:AC013603.17</host_gene>\n",
    "<host_intron></host_intron>\n",
    "<host_position>1001..1115</host_position>\n",
    "<organization>Mono</organization>\n",
    "<mod_type>H/ACA</mod_type>\n",
    "<mod_rna></mod_rna>\n",
    "<mod_site></mod_site>\n",
    "<duplex_region></duplex_region>\n",
    "<accession></accession>\n",
    "<evidence></evidence>\n",
    "<note></note>\n",
    "<sequence>\n",
    "TTGATAAGAAAGAAGTTCTGATCCCAGTGTGCAATGGCTGCAAACAGCAGCTTTCTTGGT\n",
    "GGTGTACATGGCCTGTTTCTTGTATGGGTTGCTCTAAGGGTCCTTGGAGACAGGC\n",
    "</sequence>\n",
    "</snoopy>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "data = {}\n",
    "for xml in glob.glob(\"./snOPY/snoRNAs/*.xml\"):\n",
    "    try:\n",
    "        tree = ET.parse(xml)\n",
    "        root = tree.getroot()\n",
    "    except ET.ParseError, e:\n",
    "        print \"Cannot parse: \", xml, e\n",
    "    tmp = {}\n",
    "    for child in root:\n",
    "        if child.tag == \"box\":\n",
    "            if child.text:\n",
    "                tmp[child.tag + \"_\" + child.attrib['type']] = child.text\n",
    "            else:\n",
    "                tmp[child.tag] = np.nan\n",
    "        else:\n",
    "            if child.text:\n",
    "                tmp[child.tag] = child.text.replace(\"\\n\", \"\")\n",
    "            else:\n",
    "                tmp[child.tag] = np.nan\n",
    "    data[root.attrib[\"id\"]] = tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_or_contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>snor_id</th>\n",
       "      <th>mod_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>sequence</th>\n",
       "      <th>box_d</th>\n",
       "      <th>box_c</th>\n",
       "      <th>box_h</th>\n",
       "      <th>box_aca</th>\n",
       "      <th>alias</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>mod_site</th>\n",
       "      <th>host_gene</th>\n",
       "      <th>host_id</th>\n",
       "      <th>organization</th>\n",
       "      <th>organism</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300000</th>\n",
       "      <td>  8</td>\n",
       "      <td>  33517082</td>\n",
       "      <td>  33517197</td>\n",
       "      <td> SNORA70BL3_sapiens300000</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> +</td>\n",
       "      <td> TTGATAAGAAAGAAGTTCTGATCCCAGTGTGCAATGGCTGCAAACA...</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td>   4..9</td>\n",
       "      <td> 110..112</td>\n",
       "      <td> NaN</td>\n",
       "      <td> SNORA70BL3</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td> Mono:286:AC013603.17</td>\n",
       "      <td> Homo_sapiens100000</td>\n",
       "      <td> Mono</td>\n",
       "      <td> Homo_sapiens</td>\n",
       "      <td> NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300001</th>\n",
       "      <td> 12</td>\n",
       "      <td>  55541425</td>\n",
       "      <td>  55541559</td>\n",
       "      <td>  SNORA48L6_sapiens300001</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> +</td>\n",
       "      <td> TGTTCCCGATCTGGTAGAGTAGCATCTGGTTGGTGGTGACCATCTA...</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 16..21</td>\n",
       "      <td> 129..131</td>\n",
       "      <td> NaN</td>\n",
       "      <td>  SNORA48L6</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td> Mono:245:AC121758.14</td>\n",
       "      <td> Homo_sapiens100001</td>\n",
       "      <td> Mono</td>\n",
       "      <td> Homo_sapiens</td>\n",
       "      <td> NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300002</th>\n",
       "      <td>  2</td>\n",
       "      <td> 139985463</td>\n",
       "      <td> 139985595</td>\n",
       "      <td>  SNORA72L7_sapiens300002</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> +</td>\n",
       "      <td> CTGTGAATATTCTCTGTGTTCTGATTATGTGAGAGCCAGGACAGGC...</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 59..64</td>\n",
       "      <td>   59..61</td>\n",
       "      <td> NaN</td>\n",
       "      <td>  SNORA72L7</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td>   Mono:45:AC016710.7</td>\n",
       "      <td> Homo_sapiens100002</td>\n",
       "      <td> Mono</td>\n",
       "      <td> Homo_sapiens</td>\n",
       "      <td> NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300003</th>\n",
       "      <td>  7</td>\n",
       "      <td>  12706907</td>\n",
       "      <td>  12707039</td>\n",
       "      <td>  SNORA64L2_sapiens300003</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "      <td> ACACTCTCAGCTATGTATAGTTACACCTGGCTTCACTTGTGTGACT...</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 65..70</td>\n",
       "      <td> 127..129</td>\n",
       "      <td> NaN</td>\n",
       "      <td>  SNORA64L2</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td>  Mono:182:AC011891.3</td>\n",
       "      <td> Homo_sapiens100003</td>\n",
       "      <td> Mono</td>\n",
       "      <td> Homo_sapiens</td>\n",
       "      <td> NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300004</th>\n",
       "      <td> 19</td>\n",
       "      <td>   9791642</td>\n",
       "      <td>   9791770</td>\n",
       "      <td>  SNORA70L2_sapiens300004</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> +</td>\n",
       "      <td> CTGAATTGGCAAGTCCACCAAGAAACTATAAAAATGTGTGTGTGCA...</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 23..28</td>\n",
       "      <td> 123..125</td>\n",
       "      <td> NaN</td>\n",
       "      <td>  SNORA70L2</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td>  Mono:137:AC008752.6</td>\n",
       "      <td> Homo_sapiens100004</td>\n",
       "      <td> Mono</td>\n",
       "      <td> Homo_sapiens</td>\n",
       "      <td> NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   chrom_or_contig      start        end  \\\n",
       "Homo_sapiens300000               8   33517082   33517197   \n",
       "Homo_sapiens300001              12   55541425   55541559   \n",
       "Homo_sapiens300002               2  139985463  139985595   \n",
       "Homo_sapiens300003               7   12706907   12707039   \n",
       "Homo_sapiens300004              19    9791642    9791770   \n",
       "\n",
       "                                     snor_id mod_type strand  \\\n",
       "Homo_sapiens300000  SNORA70BL3_sapiens300000    H/ACA      +   \n",
       "Homo_sapiens300001   SNORA48L6_sapiens300001    H/ACA      +   \n",
       "Homo_sapiens300002   SNORA72L7_sapiens300002    H/ACA      +   \n",
       "Homo_sapiens300003   SNORA64L2_sapiens300003    H/ACA      -   \n",
       "Homo_sapiens300004   SNORA70L2_sapiens300004    H/ACA      +   \n",
       "\n",
       "                                                             sequence box_d  \\\n",
       "Homo_sapiens300000  TTGATAAGAAAGAAGTTCTGATCCCAGTGTGCAATGGCTGCAAACA...   NaN   \n",
       "Homo_sapiens300001  TGTTCCCGATCTGGTAGAGTAGCATCTGGTTGGTGGTGACCATCTA...   NaN   \n",
       "Homo_sapiens300002  CTGTGAATATTCTCTGTGTTCTGATTATGTGAGAGCCAGGACAGGC...   NaN   \n",
       "Homo_sapiens300003  ACACTCTCAGCTATGTATAGTTACACCTGGCTTCACTTGTGTGACT...   NaN   \n",
       "Homo_sapiens300004  CTGAATTGGCAAGTCCACCAAGAAACTATAAAAATGTGTGTGTGCA...   NaN   \n",
       "\n",
       "                   box_c   box_h   box_aca alias   gene_name accession  \\\n",
       "Homo_sapiens300000   NaN    4..9  110..112   NaN  SNORA70BL3       NaN   \n",
       "Homo_sapiens300001   NaN  16..21  129..131   NaN   SNORA48L6       NaN   \n",
       "Homo_sapiens300002   NaN  59..64    59..61   NaN   SNORA72L7       NaN   \n",
       "Homo_sapiens300003   NaN  65..70  127..129   NaN   SNORA64L2       NaN   \n",
       "Homo_sapiens300004   NaN  23..28  123..125   NaN   SNORA70L2       NaN   \n",
       "\n",
       "                   mod_site             host_gene             host_id  \\\n",
       "Homo_sapiens300000      NaN  Mono:286:AC013603.17  Homo_sapiens100000   \n",
       "Homo_sapiens300001      NaN  Mono:245:AC121758.14  Homo_sapiens100001   \n",
       "Homo_sapiens300002      NaN    Mono:45:AC016710.7  Homo_sapiens100002   \n",
       "Homo_sapiens300003      NaN   Mono:182:AC011891.3  Homo_sapiens100003   \n",
       "Homo_sapiens300004      NaN   Mono:137:AC008752.6  Homo_sapiens100004   \n",
       "\n",
       "                   organization      organism note  \n",
       "Homo_sapiens300000         Mono  Homo_sapiens  NaN  \n",
       "Homo_sapiens300001         Mono  Homo_sapiens  NaN  \n",
       "Homo_sapiens300002         Mono  Homo_sapiens  NaN  \n",
       "Homo_sapiens300003         Mono  Homo_sapiens  NaN  \n",
       "Homo_sapiens300004         Mono  Homo_sapiens  NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(data).T\n",
    "df[\"id\"] = [i.split(\"_\")[-1] for i in df.index.tolist()]\n",
    "df = df.dropna(subset=[\"mod_type\", \"genome_position\"])\n",
    "df[\"start\"] = [int(i.split(\"..\")[0]) - 1 for i in df[\"genome_position\"]]\n",
    "df[\"end\"] = [int(i.split(\"..\")[1]) for i in df[\"genome_position\"]]\n",
    "df[\"snor_id\"] = df.gene_name + \"_\" + df.id\n",
    "df = df[[\"chrom_or_contig\",\n",
    "    \"start\",\n",
    "    \"end\",\n",
    "    \"snor_id\",\n",
    "    \"mod_type\",\n",
    "    \"strand\",\n",
    "    \"sequence\",\n",
    "    \"box_d\",\n",
    "    \"box_c\",\n",
    "    \"box_h\",\n",
    "    \"box_aca\",\n",
    "    \"alias\",\n",
    "    \"gene_name\",\n",
    "    \"accession\",\n",
    "    \"mod_site\",\n",
    "    \"host_gene\",\n",
    "    \"host_id\",\n",
    "    \"organization\",\n",
    "    \"organism\",\n",
    "    \"note\"]]\n",
    "df.mod_type =df.mod_type.str.replace(\" Pu\", \"\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Write columns help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "desc_map = OrderedDict(((\"Chromosome\", \"Chromosome on which gene is located\"),\n",
    "            (\"Start\", \"0-based start coordinate\"),\n",
    "            (\"End\", \"1-based end coordinate\"),\n",
    "            (\"snoRNA ID\", \"ID for snoRNA\"),\n",
    "            (\"Type\", \"Type of snoRNA (C/D or H/ACA)\"),\n",
    "            (\"Strand\", \"Strand on which gene is located\"),\n",
    "            (\"Sequence\", \"Sequence of snoRNA\"),\n",
    "            (\"D-boxes\", \"1-based coordinates snoRNA D-boxes in form of \"+\n",
    "                       \"D-box_start..D-box_end or D'-box_start..D'-box_end,D-box_start..D-box_end \"+\n",
    "                       \"eg. 60..63 or 30..33,60..63. It must be 4 nucleotides long.\"),\n",
    "            (\"C-boxes\", \"Same as above (not that prime boxes are reversed in positions) \"+\n",
    "                       \"with exception of the length constraint\"),\n",
    "            (\"H-box\", \"1-based position of the H box in H/ACA box snoRNAs in form of start..end\"),\n",
    "            (\"ACA-box\", \"1-based position of the ACA box in H/ACA box snoRNA in form of start..end\"),\n",
    "            (\"Alias\", \"Alias/Alternative name for snoRNA\"),\n",
    "            (\"Gene name\", \"Name of the snoRNA gene\"),\n",
    "            (\"Accession\", \"Accession (NCBI, RefSeq etc) for snoRNA\"),\n",
    "            (\"Modification sites\", \"Coma separated list of modified sites in form of \"+\n",
    "                                  \"ModifiedRNA:NucleotidePosition eg. 28S:U46,18S:G52\"),\n",
    "            (\"Host gene\", \"Host gene name if any\"),\n",
    "            (\"Host ID\", \"Host locus ID\"),\n",
    "            (\"Organization\", \"Organisation of the gene eg. intronic\"),\n",
    "            (\"Organism\", \"Species\"),\n",
    "            (\"Note\", \"Additional info about snoRNA\"))\n",
    ")\n",
    "with open(\"Columns_description.tab\", \"w\") as col_desc:\n",
    "    counter = 1\n",
    "    for col, desc in desc_map.iteritems():\n",
    "        col_desc.write(\"%i. %s\\t%s\\n\" % (counter,\n",
    "                                         col,\n",
    "                                         desc))\n",
    "        counter += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Homo sapiens"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####Lift over hg18 coordinates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The procedure of this is little bit tedious. First we have to check which assembly it is. Fortunately we have sequences of snoRNAs so we can use local alignment to check it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "to_save = [\"chrom_or_contig\",\n",
    "           \"start\",\n",
    "           \"end\",\n",
    "           \"snor_id\",\n",
    "           \"mod_type\",\n",
    "           \"strand\",\n",
    "           \"sequence\"]\n",
    "homo_sapiens = df[(df.organism == \"Homo_sapiens\")]\n",
    "homo_sapiens.chrom_or_contig = \"chr\" + homo_sapiens.chrom_or_contig\n",
    "homo_sapiens[to_save].to_csv(\"human/hsa_snoRNAs.bed\",\n",
    "                             header=None,\n",
    "                             index=None,\n",
    "                             sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After the data preparation we extract sequences from hg19 and hg18 and append them to the previously generated file. In the end we read generated file into variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " Warning:  chr12:132515768-132515905 out of range\n",
      " Warning:  chr16:89627837-89627909 out of range\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "rg-extract-sequences.py \\\n",
    "    --input human/hsa_snoRNAs.bed \\\n",
    "    --output human/hsa_snoRNAs.bed.hg19 \\\n",
    "    --genome-dir human/hg19/\n",
    "rg-extract-sequences.py \\\n",
    "    --input human/hsa_snoRNAs.bed.hg19 \\\n",
    "    --output human/hsa_snoRNAs.bed.hg19.hg18 \\\n",
    "    --genome-dir /import/bc2/home/zavolan/GROUP/DB/UCSC/hg18/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for_assembly_human = pd.read_table(\"human/hsa_snoRNAs.bed.hg19.hg18\",\n",
    "                             header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td> chrY</td>\n",
       "      <td> 26802917</td>\n",
       "      <td> 26803056</td>\n",
       "      <td> SNORA70BL6_sapiens300100</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> +</td>\n",
       "      <td> AGATCAGCTTTTCAAAGGTTTACTGCATTTTACCCTCCCTGATGGG...</td>\n",
       "      <td> attgtgggttgccagggtcacacatgcaattactgatttactgggt...</td>\n",
       "      <td> agatcagcttttcaaaggtttaCTGCATTTTACCCTCCCTGATGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td> chrY</td>\n",
       "      <td> 16759521</td>\n",
       "      <td> 16759653</td>\n",
       "      <td>  SNORA20L4_sapiens300149</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "      <td> CTTCCCATTTATTTGCTGGTATAATGATACAATGATATGAGCAGTT...</td>\n",
       "      <td> AATGACAACTGATTAATTTTGTTAATAGTAATGGAAGGTTTTCTTT...</td>\n",
       "      <td> CTTCCCATTTATTTGCTGGTATAATGATACAATGATATGAGCAGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td> chrY</td>\n",
       "      <td> 23978633</td>\n",
       "      <td> 23978772</td>\n",
       "      <td> SNORA70BL5_sapiens300155</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "      <td> AGATCAGCTTTTCAAAGGTTTACTGCATTTTACCCTCCCTGATGGG...</td>\n",
       "      <td> TTGTGAAAAAGGCTGGCTGTCAAAACCCACTTCAAGACCCTAAAAG...</td>\n",
       "      <td> agatcagcttttcaaaggtttaCTGCATTTTACCCTCCCTGATGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td> chrX</td>\n",
       "      <td> 86288391</td>\n",
       "      <td> 86288463</td>\n",
       "      <td> SNORD45BL2_sapiens300006</td>\n",
       "      <td>   C/D</td>\n",
       "      <td> +</td>\n",
       "      <td> GGTCAATGATGTAATGGCATGTATTAGCTGAATCCAAAGTTGAAGT...</td>\n",
       "      <td> agttcccagacagggtcgcggcggggcagaggcgctccccacatcc...</td>\n",
       "      <td> ggTCAATGATGTAATGGCATGTATTAGCTGAATCCAAAGTTGAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td> chrX</td>\n",
       "      <td>  3460157</td>\n",
       "      <td>  3460292</td>\n",
       "      <td>  SNORA48L4_sapiens300050</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "      <td> TTTCCCTGACCTGGGTAGAGTGGCATCCAGTTGGTGGTGCCCATCT...</td>\n",
       "      <td> ATAGTCTCTGCACCCATGAAAAGTTCTTGCCCACATGTCCGATCAG...</td>\n",
       "      <td> TTTCCCTGACCTGGGTAGAGTGGCATCCAGTTGGTGGTGCCCATCT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0         1         2                         3      4  5  \\\n",
       "0  chrY  26802917  26803056  SNORA70BL6_sapiens300100  H/ACA  +   \n",
       "1  chrY  16759521  16759653   SNORA20L4_sapiens300149  H/ACA  -   \n",
       "2  chrY  23978633  23978772  SNORA70BL5_sapiens300155  H/ACA  -   \n",
       "3  chrX  86288391  86288463  SNORD45BL2_sapiens300006    C/D  +   \n",
       "4  chrX   3460157   3460292   SNORA48L4_sapiens300050  H/ACA  -   \n",
       "\n",
       "                                                   6  \\\n",
       "0  AGATCAGCTTTTCAAAGGTTTACTGCATTTTACCCTCCCTGATGGG...   \n",
       "1  CTTCCCATTTATTTGCTGGTATAATGATACAATGATATGAGCAGTT...   \n",
       "2  AGATCAGCTTTTCAAAGGTTTACTGCATTTTACCCTCCCTGATGGG...   \n",
       "3  GGTCAATGATGTAATGGCATGTATTAGCTGAATCCAAAGTTGAAGT...   \n",
       "4  TTTCCCTGACCTGGGTAGAGTGGCATCCAGTTGGTGGTGCCCATCT...   \n",
       "\n",
       "                                                   7  \\\n",
       "0  attgtgggttgccagggtcacacatgcaattactgatttactgggt...   \n",
       "1  AATGACAACTGATTAATTTTGTTAATAGTAATGGAAGGTTTTCTTT...   \n",
       "2  TTGTGAAAAAGGCTGGCTGTCAAAACCCACTTCAAGACCCTAAAAG...   \n",
       "3  agttcccagacagggtcgcggcggggcagaggcgctccccacatcc...   \n",
       "4  ATAGTCTCTGCACCCATGAAAAGTTCTTGCCCACATGTCCGATCAG...   \n",
       "\n",
       "                                                   8  \n",
       "0  agatcagcttttcaaaggtttaCTGCATTTTACCCTCCCTGATGGG...  \n",
       "1  CTTCCCATTTATTTGCTGGTATAATGATACAATGATATGAGCAGTT...  \n",
       "2  agatcagcttttcaaaggtttaCTGCATTTTACCCTCCCTGATGGG...  \n",
       "3  ggTCAATGATGTAATGGCATGTATTAGCTGAATCCAAAGTTGAAGT...  \n",
       "4  TTTCCCTGACCTGGGTAGAGTGGCATCCAGTTGGTGGTGCCCATCT...  \n",
       "\n",
       "[5 rows x 9 columns]"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for_assembly_human.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#####Append assembly"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we check from which assembly the sequence is generated and with this we can assign the assembly to snoRNAs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for_assembly_human = apped_assembly(for_assembly_human)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>hg19score</th>\n",
       "      <th>hg18score</th>\n",
       "      <th>assembly</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td> chrY</td>\n",
       "      <td> 26802917</td>\n",
       "      <td> 26803056</td>\n",
       "      <td> SNORA70BL6_sapiens300100</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> +</td>\n",
       "      <td> AGATCAGCTTTTCAAAGGTTTACTGCATTTTACCCTCCCTGATGGG...</td>\n",
       "      <td> attgtgggttgccagggtcacacatgcaattactgatttactgggt...</td>\n",
       "      <td> agatcagcttttcaaaggtttaCTGCATTTTACCCTCCCTGATGGG...</td>\n",
       "      <td> 15</td>\n",
       "      <td> 278</td>\n",
       "      <td> hg18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td> chrY</td>\n",
       "      <td> 16759521</td>\n",
       "      <td> 16759653</td>\n",
       "      <td>  SNORA20L4_sapiens300149</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "      <td> CTTCCCATTTATTTGCTGGTATAATGATACAATGATATGAGCAGTT...</td>\n",
       "      <td> AATGACAACTGATTAATTTTGTTAATAGTAATGGAAGGTTTTCTTT...</td>\n",
       "      <td> CTTCCCATTTATTTGCTGGTATAATGATACAATGATATGAGCAGTT...</td>\n",
       "      <td> 15</td>\n",
       "      <td> 264</td>\n",
       "      <td> hg18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td> chrY</td>\n",
       "      <td> 23978633</td>\n",
       "      <td> 23978772</td>\n",
       "      <td> SNORA70BL5_sapiens300155</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "      <td> AGATCAGCTTTTCAAAGGTTTACTGCATTTTACCCTCCCTGATGGG...</td>\n",
       "      <td> TTGTGAAAAAGGCTGGCTGTCAAAACCCACTTCAAGACCCTAAAAG...</td>\n",
       "      <td> agatcagcttttcaaaggtttaCTGCATTTTACCCTCCCTGATGGG...</td>\n",
       "      <td> 14</td>\n",
       "      <td> 278</td>\n",
       "      <td> hg18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td> chrX</td>\n",
       "      <td> 86288391</td>\n",
       "      <td> 86288463</td>\n",
       "      <td> SNORD45BL2_sapiens300006</td>\n",
       "      <td>   C/D</td>\n",
       "      <td> +</td>\n",
       "      <td> GGTCAATGATGTAATGGCATGTATTAGCTGAATCCAAAGTTGAAGT...</td>\n",
       "      <td> agttcccagacagggtcgcggcggggcagaggcgctccccacatcc...</td>\n",
       "      <td> ggTCAATGATGTAATGGCATGTATTAGCTGAATCCAAAGTTGAAGT...</td>\n",
       "      <td>  8</td>\n",
       "      <td> 144</td>\n",
       "      <td> hg18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td> chrX</td>\n",
       "      <td>  3460157</td>\n",
       "      <td>  3460292</td>\n",
       "      <td>  SNORA48L4_sapiens300050</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "      <td> TTTCCCTGACCTGGGTAGAGTGGCATCCAGTTGGTGGTGCCCATCT...</td>\n",
       "      <td> ATAGTCTCTGCACCCATGAAAAGTTCTTGCCCACATGTCCGATCAG...</td>\n",
       "      <td> TTTCCCTGACCTGGGTAGAGTGGCATCCAGTTGGTGGTGCCCATCT...</td>\n",
       "      <td> 18</td>\n",
       "      <td> 270</td>\n",
       "      <td> hg18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0         1         2                         3      4  5  \\\n",
       "0  chrY  26802917  26803056  SNORA70BL6_sapiens300100  H/ACA  +   \n",
       "1  chrY  16759521  16759653   SNORA20L4_sapiens300149  H/ACA  -   \n",
       "2  chrY  23978633  23978772  SNORA70BL5_sapiens300155  H/ACA  -   \n",
       "3  chrX  86288391  86288463  SNORD45BL2_sapiens300006    C/D  +   \n",
       "4  chrX   3460157   3460292   SNORA48L4_sapiens300050  H/ACA  -   \n",
       "\n",
       "                                                   6  \\\n",
       "0  AGATCAGCTTTTCAAAGGTTTACTGCATTTTACCCTCCCTGATGGG...   \n",
       "1  CTTCCCATTTATTTGCTGGTATAATGATACAATGATATGAGCAGTT...   \n",
       "2  AGATCAGCTTTTCAAAGGTTTACTGCATTTTACCCTCCCTGATGGG...   \n",
       "3  GGTCAATGATGTAATGGCATGTATTAGCTGAATCCAAAGTTGAAGT...   \n",
       "4  TTTCCCTGACCTGGGTAGAGTGGCATCCAGTTGGTGGTGCCCATCT...   \n",
       "\n",
       "                                                   7  \\\n",
       "0  attgtgggttgccagggtcacacatgcaattactgatttactgggt...   \n",
       "1  AATGACAACTGATTAATTTTGTTAATAGTAATGGAAGGTTTTCTTT...   \n",
       "2  TTGTGAAAAAGGCTGGCTGTCAAAACCCACTTCAAGACCCTAAAAG...   \n",
       "3  agttcccagacagggtcgcggcggggcagaggcgctccccacatcc...   \n",
       "4  ATAGTCTCTGCACCCATGAAAAGTTCTTGCCCACATGTCCGATCAG...   \n",
       "\n",
       "                                                   8  hg19score  hg18score  \\\n",
       "0  agatcagcttttcaaaggtttaCTGCATTTTACCCTCCCTGATGGG...         15        278   \n",
       "1  CTTCCCATTTATTTGCTGGTATAATGATACAATGATATGAGCAGTT...         15        264   \n",
       "2  agatcagcttttcaaaggtttaCTGCATTTTACCCTCCCTGATGGG...         14        278   \n",
       "3  ggTCAATGATGTAATGGCATGTATTAGCTGAATCCAAAGTTGAAGT...          8        144   \n",
       "4  TTTCCCTGACCTGGGTAGAGTGGCATCCAGTTGGTGGTGCCCATCT...         18        270   \n",
       "\n",
       "  assembly  \n",
       "0     hg18  \n",
       "1     hg18  \n",
       "2     hg18  \n",
       "3     hg18  \n",
       "4     hg18  \n",
       "\n",
       "[5 rows x 12 columns]"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for_assembly_human.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A quick look on the counts for each assembly assures us that most of the snoRNAs are derived from hg18. In this case we need to lift them over."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ambiguous</th>\n",
       "      <td>   1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hg18</th>\n",
       "      <td> 515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hg19</th>\n",
       "      <td> 245</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             0\n",
       "assembly      \n",
       "ambiguous    1\n",
       "hg18       515\n",
       "hg19       245\n",
       "\n",
       "[3 rows x 1 columns]"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(for_assembly_human.groupby(\"assembly\").size())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#####Actual lift over"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we are ready to save the data and lift over coordinates from that need to be lifted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for_assembly_human[for_assembly_human.assembly=='hg18'].ix[:,:5].to_csv(\n",
    "        \"human/hsa_snoRNAs_to_lift_over.bed\",\n",
    "        sep=\"\\t\",\n",
    "        header=None,\n",
    "        index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@ 2015-03-13 15:24:31: Read chain_file:  genome_conversion_files/hg18ToHg19.over.chain.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "CrossMap.py bed genome_conversion_files/hg18ToHg19.over.chain.gz \\\n",
    "                human/hsa_snoRNAs_to_lift_over.bed |\\\n",
    "                cut -f 8,9,10,11,12,13,14 > \\\n",
    "                human/hsa_snoRNAs_to_lift_over.bed.lifted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "human_lifted = pd.read_table(\"human/hsa_snoRNAs_to_lift_over.bed.lifted\",\n",
    "                             header=None)\n",
    "human_lifted.index = [\"Homo_%s\" % i.split(\"_\")[-1] for i in human_lifted[3]]\n",
    "for_assembly_human.index = [\"Homo_%s\" % i.split(\"_\")[-1] \n",
    "                                for i in for_assembly_human[3]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quick look on all the datasets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300100</th>\n",
       "      <td> chrY</td>\n",
       "      <td> 28393529</td>\n",
       "      <td> 28393668</td>\n",
       "      <td> SNORA70BL6_sapiens300100</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300149</th>\n",
       "      <td> chrY</td>\n",
       "      <td> 18250127</td>\n",
       "      <td> 18250259</td>\n",
       "      <td>  SNORA20L4_sapiens300149</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300155</th>\n",
       "      <td> chrY</td>\n",
       "      <td> 25569245</td>\n",
       "      <td> 25569384</td>\n",
       "      <td> SNORA70BL5_sapiens300155</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300006</th>\n",
       "      <td> chrX</td>\n",
       "      <td> 86401735</td>\n",
       "      <td> 86401807</td>\n",
       "      <td> SNORD45BL2_sapiens300006</td>\n",
       "      <td>   C/D</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300050</th>\n",
       "      <td> chrX</td>\n",
       "      <td>  3450157</td>\n",
       "      <td>  3450292</td>\n",
       "      <td>  SNORA48L4_sapiens300050</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       0         1         2                         3      4  \\\n",
       "Homo_sapiens300100  chrY  28393529  28393668  SNORA70BL6_sapiens300100  H/ACA   \n",
       "Homo_sapiens300149  chrY  18250127  18250259   SNORA20L4_sapiens300149  H/ACA   \n",
       "Homo_sapiens300155  chrY  25569245  25569384  SNORA70BL5_sapiens300155  H/ACA   \n",
       "Homo_sapiens300006  chrX  86401735  86401807  SNORD45BL2_sapiens300006    C/D   \n",
       "Homo_sapiens300050  chrX   3450157   3450292   SNORA48L4_sapiens300050  H/ACA   \n",
       "\n",
       "                    5  \n",
       "Homo_sapiens300100  +  \n",
       "Homo_sapiens300149  -  \n",
       "Homo_sapiens300155  -  \n",
       "Homo_sapiens300006  +  \n",
       "Homo_sapiens300050  -  \n",
       "\n",
       "[5 rows x 6 columns]"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_lifted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>hg19score</th>\n",
       "      <th>hg18score</th>\n",
       "      <th>assembly</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300100</th>\n",
       "      <td> chrY</td>\n",
       "      <td> 26802917</td>\n",
       "      <td> 26803056</td>\n",
       "      <td> SNORA70BL6_sapiens300100</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> +</td>\n",
       "      <td> AGATCAGCTTTTCAAAGGTTTACTGCATTTTACCCTCCCTGATGGG...</td>\n",
       "      <td> attgtgggttgccagggtcacacatgcaattactgatttactgggt...</td>\n",
       "      <td> agatcagcttttcaaaggtttaCTGCATTTTACCCTCCCTGATGGG...</td>\n",
       "      <td> 15</td>\n",
       "      <td> 278</td>\n",
       "      <td> hg18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300149</th>\n",
       "      <td> chrY</td>\n",
       "      <td> 16759521</td>\n",
       "      <td> 16759653</td>\n",
       "      <td>  SNORA20L4_sapiens300149</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "      <td> CTTCCCATTTATTTGCTGGTATAATGATACAATGATATGAGCAGTT...</td>\n",
       "      <td> AATGACAACTGATTAATTTTGTTAATAGTAATGGAAGGTTTTCTTT...</td>\n",
       "      <td> CTTCCCATTTATTTGCTGGTATAATGATACAATGATATGAGCAGTT...</td>\n",
       "      <td> 15</td>\n",
       "      <td> 264</td>\n",
       "      <td> hg18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300155</th>\n",
       "      <td> chrY</td>\n",
       "      <td> 23978633</td>\n",
       "      <td> 23978772</td>\n",
       "      <td> SNORA70BL5_sapiens300155</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "      <td> AGATCAGCTTTTCAAAGGTTTACTGCATTTTACCCTCCCTGATGGG...</td>\n",
       "      <td> TTGTGAAAAAGGCTGGCTGTCAAAACCCACTTCAAGACCCTAAAAG...</td>\n",
       "      <td> agatcagcttttcaaaggtttaCTGCATTTTACCCTCCCTGATGGG...</td>\n",
       "      <td> 14</td>\n",
       "      <td> 278</td>\n",
       "      <td> hg18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300006</th>\n",
       "      <td> chrX</td>\n",
       "      <td> 86288391</td>\n",
       "      <td> 86288463</td>\n",
       "      <td> SNORD45BL2_sapiens300006</td>\n",
       "      <td>   C/D</td>\n",
       "      <td> +</td>\n",
       "      <td> GGTCAATGATGTAATGGCATGTATTAGCTGAATCCAAAGTTGAAGT...</td>\n",
       "      <td> agttcccagacagggtcgcggcggggcagaggcgctccccacatcc...</td>\n",
       "      <td> ggTCAATGATGTAATGGCATGTATTAGCTGAATCCAAAGTTGAAGT...</td>\n",
       "      <td>  8</td>\n",
       "      <td> 144</td>\n",
       "      <td> hg18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300050</th>\n",
       "      <td> chrX</td>\n",
       "      <td>  3460157</td>\n",
       "      <td>  3460292</td>\n",
       "      <td>  SNORA48L4_sapiens300050</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "      <td> TTTCCCTGACCTGGGTAGAGTGGCATCCAGTTGGTGGTGCCCATCT...</td>\n",
       "      <td> ATAGTCTCTGCACCCATGAAAAGTTCTTGCCCACATGTCCGATCAG...</td>\n",
       "      <td> TTTCCCTGACCTGGGTAGAGTGGCATCCAGTTGGTGGTGCCCATCT...</td>\n",
       "      <td> 18</td>\n",
       "      <td> 270</td>\n",
       "      <td> hg18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       0         1         2                         3      4  \\\n",
       "Homo_sapiens300100  chrY  26802917  26803056  SNORA70BL6_sapiens300100  H/ACA   \n",
       "Homo_sapiens300149  chrY  16759521  16759653   SNORA20L4_sapiens300149  H/ACA   \n",
       "Homo_sapiens300155  chrY  23978633  23978772  SNORA70BL5_sapiens300155  H/ACA   \n",
       "Homo_sapiens300006  chrX  86288391  86288463  SNORD45BL2_sapiens300006    C/D   \n",
       "Homo_sapiens300050  chrX   3460157   3460292   SNORA48L4_sapiens300050  H/ACA   \n",
       "\n",
       "                    5                                                  6  \\\n",
       "Homo_sapiens300100  +  AGATCAGCTTTTCAAAGGTTTACTGCATTTTACCCTCCCTGATGGG...   \n",
       "Homo_sapiens300149  -  CTTCCCATTTATTTGCTGGTATAATGATACAATGATATGAGCAGTT...   \n",
       "Homo_sapiens300155  -  AGATCAGCTTTTCAAAGGTTTACTGCATTTTACCCTCCCTGATGGG...   \n",
       "Homo_sapiens300006  +  GGTCAATGATGTAATGGCATGTATTAGCTGAATCCAAAGTTGAAGT...   \n",
       "Homo_sapiens300050  -  TTTCCCTGACCTGGGTAGAGTGGCATCCAGTTGGTGGTGCCCATCT...   \n",
       "\n",
       "                                                                    7  \\\n",
       "Homo_sapiens300100  attgtgggttgccagggtcacacatgcaattactgatttactgggt...   \n",
       "Homo_sapiens300149  AATGACAACTGATTAATTTTGTTAATAGTAATGGAAGGTTTTCTTT...   \n",
       "Homo_sapiens300155  TTGTGAAAAAGGCTGGCTGTCAAAACCCACTTCAAGACCCTAAAAG...   \n",
       "Homo_sapiens300006  agttcccagacagggtcgcggcggggcagaggcgctccccacatcc...   \n",
       "Homo_sapiens300050  ATAGTCTCTGCACCCATGAAAAGTTCTTGCCCACATGTCCGATCAG...   \n",
       "\n",
       "                                                                    8  \\\n",
       "Homo_sapiens300100  agatcagcttttcaaaggtttaCTGCATTTTACCCTCCCTGATGGG...   \n",
       "Homo_sapiens300149  CTTCCCATTTATTTGCTGGTATAATGATACAATGATATGAGCAGTT...   \n",
       "Homo_sapiens300155  agatcagcttttcaaaggtttaCTGCATTTTACCCTCCCTGATGGG...   \n",
       "Homo_sapiens300006  ggTCAATGATGTAATGGCATGTATTAGCTGAATCCAAAGTTGAAGT...   \n",
       "Homo_sapiens300050  TTTCCCTGACCTGGGTAGAGTGGCATCCAGTTGGTGGTGCCCATCT...   \n",
       "\n",
       "                    hg19score  hg18score assembly  \n",
       "Homo_sapiens300100         15        278     hg18  \n",
       "Homo_sapiens300149         15        264     hg18  \n",
       "Homo_sapiens300155         14        278     hg18  \n",
       "Homo_sapiens300006          8        144     hg18  \n",
       "Homo_sapiens300050         18        270     hg18  \n",
       "\n",
       "[5 rows x 12 columns]"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for_assembly_human.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_or_contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>snor_id</th>\n",
       "      <th>mod_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>sequence</th>\n",
       "      <th>box_d</th>\n",
       "      <th>box_c</th>\n",
       "      <th>box_h</th>\n",
       "      <th>box_aca</th>\n",
       "      <th>alias</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>mod_site</th>\n",
       "      <th>host_gene</th>\n",
       "      <th>host_id</th>\n",
       "      <th>organization</th>\n",
       "      <th>organism</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300000</th>\n",
       "      <td>  chr8</td>\n",
       "      <td>  33517082</td>\n",
       "      <td>  33517197</td>\n",
       "      <td> SNORA70BL3_sapiens300000</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> +</td>\n",
       "      <td> TTGATAAGAAAGAAGTTCTGATCCCAGTGTGCAATGGCTGCAAACA...</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td>   4..9</td>\n",
       "      <td> 110..112</td>\n",
       "      <td> NaN</td>\n",
       "      <td> SNORA70BL3</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td> Mono:286:AC013603.17</td>\n",
       "      <td> Homo_sapiens100000</td>\n",
       "      <td> Mono</td>\n",
       "      <td> Homo_sapiens</td>\n",
       "      <td> NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300001</th>\n",
       "      <td> chr12</td>\n",
       "      <td>  55541425</td>\n",
       "      <td>  55541559</td>\n",
       "      <td>  SNORA48L6_sapiens300001</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> +</td>\n",
       "      <td> TGTTCCCGATCTGGTAGAGTAGCATCTGGTTGGTGGTGACCATCTA...</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 16..21</td>\n",
       "      <td> 129..131</td>\n",
       "      <td> NaN</td>\n",
       "      <td>  SNORA48L6</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td> Mono:245:AC121758.14</td>\n",
       "      <td> Homo_sapiens100001</td>\n",
       "      <td> Mono</td>\n",
       "      <td> Homo_sapiens</td>\n",
       "      <td> NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300002</th>\n",
       "      <td>  chr2</td>\n",
       "      <td> 139985463</td>\n",
       "      <td> 139985595</td>\n",
       "      <td>  SNORA72L7_sapiens300002</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> +</td>\n",
       "      <td> CTGTGAATATTCTCTGTGTTCTGATTATGTGAGAGCCAGGACAGGC...</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 59..64</td>\n",
       "      <td>   59..61</td>\n",
       "      <td> NaN</td>\n",
       "      <td>  SNORA72L7</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td>   Mono:45:AC016710.7</td>\n",
       "      <td> Homo_sapiens100002</td>\n",
       "      <td> Mono</td>\n",
       "      <td> Homo_sapiens</td>\n",
       "      <td> NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300003</th>\n",
       "      <td>  chr7</td>\n",
       "      <td>  12706907</td>\n",
       "      <td>  12707039</td>\n",
       "      <td>  SNORA64L2_sapiens300003</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "      <td> ACACTCTCAGCTATGTATAGTTACACCTGGCTTCACTTGTGTGACT...</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 65..70</td>\n",
       "      <td> 127..129</td>\n",
       "      <td> NaN</td>\n",
       "      <td>  SNORA64L2</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td>  Mono:182:AC011891.3</td>\n",
       "      <td> Homo_sapiens100003</td>\n",
       "      <td> Mono</td>\n",
       "      <td> Homo_sapiens</td>\n",
       "      <td> NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300004</th>\n",
       "      <td> chr19</td>\n",
       "      <td>   9791642</td>\n",
       "      <td>   9791770</td>\n",
       "      <td>  SNORA70L2_sapiens300004</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> +</td>\n",
       "      <td> CTGAATTGGCAAGTCCACCAAGAAACTATAAAAATGTGTGTGTGCA...</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 23..28</td>\n",
       "      <td> 123..125</td>\n",
       "      <td> NaN</td>\n",
       "      <td>  SNORA70L2</td>\n",
       "      <td> NaN</td>\n",
       "      <td> NaN</td>\n",
       "      <td>  Mono:137:AC008752.6</td>\n",
       "      <td> Homo_sapiens100004</td>\n",
       "      <td> Mono</td>\n",
       "      <td> Homo_sapiens</td>\n",
       "      <td> NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   chrom_or_contig      start        end  \\\n",
       "Homo_sapiens300000            chr8   33517082   33517197   \n",
       "Homo_sapiens300001           chr12   55541425   55541559   \n",
       "Homo_sapiens300002            chr2  139985463  139985595   \n",
       "Homo_sapiens300003            chr7   12706907   12707039   \n",
       "Homo_sapiens300004           chr19    9791642    9791770   \n",
       "\n",
       "                                     snor_id mod_type strand  \\\n",
       "Homo_sapiens300000  SNORA70BL3_sapiens300000    H/ACA      +   \n",
       "Homo_sapiens300001   SNORA48L6_sapiens300001    H/ACA      +   \n",
       "Homo_sapiens300002   SNORA72L7_sapiens300002    H/ACA      +   \n",
       "Homo_sapiens300003   SNORA64L2_sapiens300003    H/ACA      -   \n",
       "Homo_sapiens300004   SNORA70L2_sapiens300004    H/ACA      +   \n",
       "\n",
       "                                                             sequence box_d  \\\n",
       "Homo_sapiens300000  TTGATAAGAAAGAAGTTCTGATCCCAGTGTGCAATGGCTGCAAACA...   NaN   \n",
       "Homo_sapiens300001  TGTTCCCGATCTGGTAGAGTAGCATCTGGTTGGTGGTGACCATCTA...   NaN   \n",
       "Homo_sapiens300002  CTGTGAATATTCTCTGTGTTCTGATTATGTGAGAGCCAGGACAGGC...   NaN   \n",
       "Homo_sapiens300003  ACACTCTCAGCTATGTATAGTTACACCTGGCTTCACTTGTGTGACT...   NaN   \n",
       "Homo_sapiens300004  CTGAATTGGCAAGTCCACCAAGAAACTATAAAAATGTGTGTGTGCA...   NaN   \n",
       "\n",
       "                   box_c   box_h   box_aca alias   gene_name accession  \\\n",
       "Homo_sapiens300000   NaN    4..9  110..112   NaN  SNORA70BL3       NaN   \n",
       "Homo_sapiens300001   NaN  16..21  129..131   NaN   SNORA48L6       NaN   \n",
       "Homo_sapiens300002   NaN  59..64    59..61   NaN   SNORA72L7       NaN   \n",
       "Homo_sapiens300003   NaN  65..70  127..129   NaN   SNORA64L2       NaN   \n",
       "Homo_sapiens300004   NaN  23..28  123..125   NaN   SNORA70L2       NaN   \n",
       "\n",
       "                   mod_site             host_gene             host_id  \\\n",
       "Homo_sapiens300000      NaN  Mono:286:AC013603.17  Homo_sapiens100000   \n",
       "Homo_sapiens300001      NaN  Mono:245:AC121758.14  Homo_sapiens100001   \n",
       "Homo_sapiens300002      NaN    Mono:45:AC016710.7  Homo_sapiens100002   \n",
       "Homo_sapiens300003      NaN   Mono:182:AC011891.3  Homo_sapiens100003   \n",
       "Homo_sapiens300004      NaN   Mono:137:AC008752.6  Homo_sapiens100004   \n",
       "\n",
       "                   organization      organism note  \n",
       "Homo_sapiens300000         Mono  Homo_sapiens  NaN  \n",
       "Homo_sapiens300001         Mono  Homo_sapiens  NaN  \n",
       "Homo_sapiens300002         Mono  Homo_sapiens  NaN  \n",
       "Homo_sapiens300003         Mono  Homo_sapiens  NaN  \n",
       "Homo_sapiens300004         Mono  Homo_sapiens  NaN  \n",
       "\n",
       "[5 rows x 20 columns]"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "homo_sapiens.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will change coordinates in the main DataFrame if they are hg18."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "homo_sapiens_with_assembly = homo_sapiens.join(for_assembly_human['assembly'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "new_chrom = []\n",
    "new_start = []\n",
    "new_end = []\n",
    "for idx, row in homo_sapiens_with_assembly.iterrows():\n",
    "    if row.assembly == \"hg19\" or row.assembly == 'ambiguous':\n",
    "        new_chrom.append(row.chrom_or_contig)\n",
    "        new_start.append(row.start)\n",
    "        new_end.append(row.end)\n",
    "    elif row.assembly == \"hg18\":\n",
    "        new_chrom.append(human_lifted.ix[idx, 0])\n",
    "        new_start.append(human_lifted.ix[idx, 1])\n",
    "        new_end.append(human_lifted.ix[idx, 2])\n",
    "    else:\n",
    "        print row\n",
    "        raise Exception(\"Do not know assembly named %s\" % row.assembly)\n",
    "homo_sapiens_with_assembly.chrom_or_contig = new_chrom\n",
    "homo_sapiens_with_assembly.start = new_start\n",
    "homo_sapiens_with_assembly.end = new_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "del homo_sapiens_with_assembly[\"assembly\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#####Check if there is any strange annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just quick look if there is anything strange in the annotations. (Both is not strange)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_or_contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>snor_id</th>\n",
       "      <th>mod_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>sequence</th>\n",
       "      <th>box_d</th>\n",
       "      <th>box_c</th>\n",
       "      <th>box_h</th>\n",
       "      <th>box_aca</th>\n",
       "      <th>alias</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>mod_site</th>\n",
       "      <th>host_gene</th>\n",
       "      <th>host_id</th>\n",
       "      <th>organization</th>\n",
       "      <th>organism</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens300786</th>\n",
       "      <td> chr12</td>\n",
       "      <td> 6619387</td>\n",
       "      <td> 6619717</td>\n",
       "      <td> SCARNA10_sapiens300786</td>\n",
       "      <td> Both</td>\n",
       "      <td> +</td>\n",
       "      <td> GCCACATGATGATATCAAGGCTGTTGTGATTCAGTTGGTTTGGCTA...</td>\n",
       "      <td> 322..325</td>\n",
       "      <td> 7..12</td>\n",
       "      <td> 131..136</td>\n",
       "      <td> NaN</td>\n",
       "      <td> U85</td>\n",
       "      <td> SCARNA10</td>\n",
       "      <td> NR_004387,NT_009759</td>\n",
       "      <td> U5:U46,U5:C45</td>\n",
       "      <td> NCAPD2</td>\n",
       "      <td> Homo_sapiens100543</td>\n",
       "      <td> Intronic</td>\n",
       "      <td> Homo_sapiens</td>\n",
       "      <td> NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   chrom_or_contig    start      end                 snor_id  \\\n",
       "Homo_sapiens300786           chr12  6619387  6619717  SCARNA10_sapiens300786   \n",
       "\n",
       "                   mod_type strand  \\\n",
       "Homo_sapiens300786     Both      +   \n",
       "\n",
       "                                                             sequence  \\\n",
       "Homo_sapiens300786  GCCACATGATGATATCAAGGCTGTTGTGATTCAGTTGGTTTGGCTA...   \n",
       "\n",
       "                       box_d  box_c     box_h box_aca alias gene_name  \\\n",
       "Homo_sapiens300786  322..325  7..12  131..136     NaN   U85  SCARNA10   \n",
       "\n",
       "                              accession       mod_site host_gene  \\\n",
       "Homo_sapiens300786  NR_004387,NT_009759  U5:U46,U5:C45    NCAPD2   \n",
       "\n",
       "                               host_id organization      organism note  \n",
       "Homo_sapiens300786  Homo_sapiens100543     Intronic  Homo_sapiens  NaN  \n",
       "\n",
       "[1 rows x 20 columns]"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "homo_sapiens_with_assembly[(homo_sapiens_with_assembly.mod_type!=\"C/D\") & \n",
    "                          (homo_sapiens_with_assembly.mod_type!=\"H/ACA\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####Save snoRNAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "homo_sapiens_cd = homo_sapiens_with_assembly[homo_sapiens_with_assembly.mod_type.str.contains(\"C/D\")]\n",
    "homo_sapiens_cd = homo_sapiens_cd.dropna(subset=[\"box_d\"])\n",
    "homo_sapiens_aca = homo_sapiens_with_assembly[homo_sapiens_with_assembly.mod_type.str.contains(\"H/ACA\")]\n",
    "homo_sapiens_aca = homo_sapiens_aca.dropna(subset=[\"box_h\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "homo_sapiens_cd.to_csv(\"human/hsa_cd_box_snoRNAs.tab\",\n",
    "                       index=False,\n",
    "                       header=False,\n",
    "                       sep=\"\\t\",\n",
    "                       na_rep=\"NaN\")\n",
    "homo_sapiens_aca.to_csv(\"human/hsa_haca_box_snoRNAs.tab\",\n",
    "                       index=False,\n",
    "                       header=False,\n",
    "                       sep=\"\\t\",\n",
    "                       na_rep=\"NaN\",\n",
    "                        encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Mus musculus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Whole procedure should be repeated for mouse too."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/import/bc2/apps/Python/2.7.5-goolf-1.4.10/lib/python2.7/site-packages/pandas-0.15.2-py2.7-linux-x86_64.egg/pandas/core/generic.py:1974: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self[name] = value\n"
     ]
    }
   ],
   "source": [
    "to_save = [\"chrom_or_contig\",\n",
    "           \"start\",\n",
    "           \"end\",\n",
    "           \"snor_id\",\n",
    "           \"mod_type\",\n",
    "           \"strand\",\n",
    "           \"sequence\"]\n",
    "mus_musculus = df[(df.organism == \"Mus_musculus\")]\n",
    "mus_musculus.chrom_or_contig = \"chr\" + mus_musculus.chrom_or_contig\n",
    "mus_musculus.mod_type = mus_musculus.mod_type.str.replace(\" Pu\", \"\")\n",
    "mus_musculus[to_save].to_csv(\"mouse/mmu_snoRNAs.bed\",\n",
    "                             header=None,\n",
    "                             index=None,\n",
    "                             sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " Warning:  chr7:149339749-149339862 out of range\n",
      " Warning:  chr7:148572393-148572523 out of range\n",
      " Warning:  chr7:148634701-148634838 out of range\n",
      " Warning:  chr7:150717298-150717425 out of range\n",
      " Warning:  chr8:131637194-131637326 out of range\n",
      " Warning:  chr8:131541789-131541921 out of range\n",
      " Warning:  chr8:129468746-129468880 out of range\n",
      "Cannot read from mouse/mm10/.fa file\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "rg-extract-sequences.py \\\n",
    "    --input mouse/mmu_snoRNAs.bed \\\n",
    "    --output mouse/mmu_snoRNAs.bed.mm10 \\\n",
    "    --genome-dir mouse/mm10/\n",
    "rg-extract-sequences.py \\\n",
    "    --input mouse/mmu_snoRNAs.bed.mm10 \\\n",
    "    --output mouse/mmu_snoRNAs.bed.mm10.mm9 \\\n",
    "    --genome-dir /import/bc2/home/zavolan/GROUP/DB/UCSC/mm9/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for_assembly_mouse = pd.read_table(\"mouse/mmu_snoRNAs.bed.mm10.mm9\",\n",
    "                             header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for_assembly_mouse = apped_mouse_assembly(for_assembly_mouse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>mm10score</th>\n",
       "      <th>mm9score</th>\n",
       "      <th>assembly</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td> chr7</td>\n",
       "      <td> 113570225</td>\n",
       "      <td> 113570361</td>\n",
       "      <td>  SNORA17_musculus300011</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> +</td>\n",
       "      <td> TCTGCCCCAGGAGCATTGCAGGTATGACTGCTGTGGCACATGTGTG...</td>\n",
       "      <td> AGAAAGGTAATGAATAAATGATGCTCATCCTAGAATGTATAATTTT...</td>\n",
       "      <td> TCTGCCCCAGGAGCATTGCAGGTATGACTGCTGTGGCACATGTGTG...</td>\n",
       "      <td> 12</td>\n",
       "      <td> 272</td>\n",
       "      <td> mm9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td> chr7</td>\n",
       "      <td> 112790851</td>\n",
       "      <td> 112790990</td>\n",
       "      <td>  SNORA43_musculus300020</td>\n",
       "      <td> H/ACA</td>\n",
       "      <td> -</td>\n",
       "      <td> GCTGTCCTAGATCAGCACCACAGACCTGGCCTAGCTGCTCAGCCCT...</td>\n",
       "      <td> CCTCTACTGAAGAAACACTATAATGTCTGCCTTCTCACTGAGAGGG...</td>\n",
       "      <td> GCTGTCCTAGATCAGCACCACAGACCTGGCCTAGCTGCTCAGCCCT...</td>\n",
       "      <td> 12</td>\n",
       "      <td> 278</td>\n",
       "      <td> mm9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td> chr7</td>\n",
       "      <td>  67004099</td>\n",
       "      <td>  67004193</td>\n",
       "      <td> SNORD116_musculus300055</td>\n",
       "      <td>   C/D</td>\n",
       "      <td> -</td>\n",
       "      <td> GGATCTATGATGATTCCCAGTCAAACATTCCTTGGAAAAGCTGAAC...</td>\n",
       "      <td> TGAGTGCTTTTACTATTATGGCTAAATGTGGGTCTCTTCTCTGAAG...</td>\n",
       "      <td> GGATCTATGATGATTCCCAGTCAAACATTCCTTGGAAAAGCTGAAC...</td>\n",
       "      <td> 14</td>\n",
       "      <td> 188</td>\n",
       "      <td> mm9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td> chr7</td>\n",
       "      <td>  67006616</td>\n",
       "      <td>  67006710</td>\n",
       "      <td> SNORD116_musculus300056</td>\n",
       "      <td>   C/D</td>\n",
       "      <td> -</td>\n",
       "      <td> GGATCTATGATGATTCCCAGTCAAACATTCCTTGGAAAAGCTGAAC...</td>\n",
       "      <td> ttctgttagcatgccttccatcatgatggactagttctgaaataag...</td>\n",
       "      <td> GGATCTATGATGATTCCCAGTCAAACATTCCTTGGAAAAGCTGAAC...</td>\n",
       "      <td> 12</td>\n",
       "      <td> 188</td>\n",
       "      <td> mm9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td> chr7</td>\n",
       "      <td>  67009125</td>\n",
       "      <td>  67009219</td>\n",
       "      <td> SNORD116_musculus300057</td>\n",
       "      <td>   C/D</td>\n",
       "      <td> -</td>\n",
       "      <td> GGATCTATGATGATTCCCAGTCAAACATTCCTTGGAAAAGCTGAAC...</td>\n",
       "      <td> GAGGGTTGTGGGATGTAGGCAGACTCACACCTAAATGAATCATGTC...</td>\n",
       "      <td> GGATCTATGATGATTCCCAGTCAAACATTCCTTGGAAAAGCTGAAC...</td>\n",
       "      <td> 16</td>\n",
       "      <td> 188</td>\n",
       "      <td> mm9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1          2                        3      4  5  \\\n",
       "0  chr7  113570225  113570361   SNORA17_musculus300011  H/ACA  +   \n",
       "1  chr7  112790851  112790990   SNORA43_musculus300020  H/ACA  -   \n",
       "2  chr7   67004099   67004193  SNORD116_musculus300055    C/D  -   \n",
       "3  chr7   67006616   67006710  SNORD116_musculus300056    C/D  -   \n",
       "4  chr7   67009125   67009219  SNORD116_musculus300057    C/D  -   \n",
       "\n",
       "                                                   6  \\\n",
       "0  TCTGCCCCAGGAGCATTGCAGGTATGACTGCTGTGGCACATGTGTG...   \n",
       "1  GCTGTCCTAGATCAGCACCACAGACCTGGCCTAGCTGCTCAGCCCT...   \n",
       "2  GGATCTATGATGATTCCCAGTCAAACATTCCTTGGAAAAGCTGAAC...   \n",
       "3  GGATCTATGATGATTCCCAGTCAAACATTCCTTGGAAAAGCTGAAC...   \n",
       "4  GGATCTATGATGATTCCCAGTCAAACATTCCTTGGAAAAGCTGAAC...   \n",
       "\n",
       "                                                   7  \\\n",
       "0  AGAAAGGTAATGAATAAATGATGCTCATCCTAGAATGTATAATTTT...   \n",
       "1  CCTCTACTGAAGAAACACTATAATGTCTGCCTTCTCACTGAGAGGG...   \n",
       "2  TGAGTGCTTTTACTATTATGGCTAAATGTGGGTCTCTTCTCTGAAG...   \n",
       "3  ttctgttagcatgccttccatcatgatggactagttctgaaataag...   \n",
       "4  GAGGGTTGTGGGATGTAGGCAGACTCACACCTAAATGAATCATGTC...   \n",
       "\n",
       "                                                   8  mm10score  mm9score  \\\n",
       "0  TCTGCCCCAGGAGCATTGCAGGTATGACTGCTGTGGCACATGTGTG...         12       272   \n",
       "1  GCTGTCCTAGATCAGCACCACAGACCTGGCCTAGCTGCTCAGCCCT...         12       278   \n",
       "2  GGATCTATGATGATTCCCAGTCAAACATTCCTTGGAAAAGCTGAAC...         14       188   \n",
       "3  GGATCTATGATGATTCCCAGTCAAACATTCCTTGGAAAAGCTGAAC...         12       188   \n",
       "4  GGATCTATGATGATTCCCAGTCAAACATTCCTTGGAAAAGCTGAAC...         16       188   \n",
       "\n",
       "  assembly  \n",
       "0      mm9  \n",
       "1      mm9  \n",
       "2      mm9  \n",
       "3      mm9  \n",
       "4      mm9  \n",
       "\n",
       "[5 rows x 12 columns]"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for_assembly_mouse.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ambiguous</th>\n",
       "      <td>   8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mm9</th>\n",
       "      <td> 965</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             0\n",
       "assembly      \n",
       "ambiguous    8\n",
       "mm9        965\n",
       "\n",
       "[2 rows x 1 columns]"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(for_assembly_mouse.groupby(\"assembly\").size())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results above shows that all the coordinates for mouse are from mm9 assembly. This is good sign. We do not need to go though all the steps of remapping - we will only lift the coordinates over."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####Lifting over"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/import/bc2/apps/Python/2.7.5-goolf-1.4.10/lib/python2.7/site-packages/IPython/kernel/__main__.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/import/bc2/apps/Python/2.7.5-goolf-1.4.10/lib/python2.7/site-packages/IPython/kernel/__main__.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "mus_musculus_cd = df[(df.organism == \"Mus_musculus\") & (df.mod_type==\"C/D\")]\n",
    "mus_musculus_cd = mus_musculus_cd.dropna(subset=[\"box_d\"])\n",
    "mus_musculus_aca = df[(df.organism == \"Mus_musculus\") & (df.mod_type==\"H/ACA\")]\n",
    "mus_musculus_aca = mus_musculus_aca.dropna(subset=[\"box_h\"])\n",
    "\n",
    "aca_for_conversion = df[(df.organism == \"Mus_musculus\") & (df.mod_type==\"H/ACA\")]\n",
    "aca_for_conversion[\"chrom_or_contig\"] = \"chr\" + aca_for_conversion.chrom_or_contig\n",
    "aca_for_conversion.iloc[:,:6].to_csv(\"mouse/mmu_mm9_haca_box_snoRNAs.bed\",\n",
    "                                    header=None,\n",
    "                                    sep=\"\\t\",\n",
    "                                    index=None)\n",
    "\n",
    "cd_for_conversion = df[(df.organism == \"Mus_musculus\") & (df.mod_type==\"C/D\")]\n",
    "cd_for_conversion[\"chrom_or_contig\"] = \"chr\" + cd_for_conversion.chrom_or_contig\n",
    "cd_for_conversion.iloc[:,:6].to_csv(\"mouse/mmu_mm9_cd_box_snoRNAs.bed\",\n",
    "                                    header=None,\n",
    "                                    sep=\"\\t\",\n",
    "                                    index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####Convert mm9 coordinates to mm10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@ 2015-04-10 15:28:14: Read chain_file:  /import/bc2/home/zavolan/gumiennr/Data/Genome/genome_conversion_files/mm9ToMm10.over.chain.gz\n",
      "End corrdinate is not an integer. skip 166888972\t166889065\tSNORD12_musculus300988\tC/D\t+\n",
      "@ 2015-04-10 15:28:16: Read chain_file:  /import/bc2/home/zavolan/gumiennr/Data/Genome/genome_conversion_files/mm9ToMm10.over.chain.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "CrossMap.py bed ~/Data/Genome/genome_conversion_files/mm9ToMm10.over.chain.gz \\\n",
    "                mouse/mmu_mm9_cd_box_snoRNAs.bed | \\\n",
    "                cut -f 8,9,10,11,12,13,14 > \\\n",
    "                mouse/mmu_mm10_cd_box_snoRNAs.bed\n",
    "CrossMap.py bed ~/Data/Genome/genome_conversion_files/mm9ToMm10.over.chain.gz \\\n",
    "                mouse/mmu_mm9_haca_box_snoRNAs.bed | \\\n",
    "                cut -f 8,9,10,11,12,13,14 > \\\n",
    "                mouse/mmu_mm10_haca_box_snoRNAs.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cd_mm9_2_mm10 = pd.read_table(\"mouse/mmu_mm10_cd_box_snoRNAs.bed\", index_col=3,\n",
    "                           header=None)\n",
    "aca_mm9_2_mm10 = pd.read_table(\"mouse/mmu_mm10_haca_box_snoRNAs.bed\", index_col=3,\n",
    "                           header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SNORD86_musculus300006</th>\n",
       "      <td> chr14</td>\n",
       "      <td>  56837070</td>\n",
       "      <td>  56837156</td>\n",
       "      <td> C/D</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNORD23_musculus300008</th>\n",
       "      <td> chr10</td>\n",
       "      <td>  21878783</td>\n",
       "      <td>  21878892</td>\n",
       "      <td> C/D</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNORND104_musculus300032</th>\n",
       "      <td> chr11</td>\n",
       "      <td> 106500992</td>\n",
       "      <td> 106501062</td>\n",
       "      <td> C/D</td>\n",
       "      <td> +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNORD116_musculus300055</th>\n",
       "      <td>  chr7</td>\n",
       "      <td>  59859213</td>\n",
       "      <td>  59859307</td>\n",
       "      <td> C/D</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNORD116_musculus300056</th>\n",
       "      <td>  chr7</td>\n",
       "      <td>  59861730</td>\n",
       "      <td>  59861824</td>\n",
       "      <td> C/D</td>\n",
       "      <td> -</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              0          1          2    4  5\n",
       "3                                                            \n",
       "SNORD86_musculus300006    chr14   56837070   56837156  C/D  -\n",
       "SNORD23_musculus300008    chr10   21878783   21878892  C/D  -\n",
       "SNORND104_musculus300032  chr11  106500992  106501062  C/D  +\n",
       "SNORD116_musculus300055    chr7   59859213   59859307  C/D  -\n",
       "SNORD116_musculus300056    chr7   59861730   59861824  C/D  -"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cd_mm9_2_mm10.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mus_musculus_cd[\"start\"] = [int(cd_mm9_2_mm10.ix[i, 1]) if i in cd_mm9_2_mm10.index\n",
    "                            else np.nan for i in mus_musculus_cd.snor_id]\n",
    "mus_musculus_cd[\"end\"] = [int(cd_mm9_2_mm10.ix[i, 2]) if i in cd_mm9_2_mm10.index\n",
    "                            else np.nan for i in mus_musculus_cd.snor_id]\n",
    "mus_musculus_cd = mus_musculus_cd.dropna(subset=[\"start\"])\n",
    "mus_musculus_cd[\"end\"] = mus_musculus_cd[\"end\"].map(int)\n",
    "mus_musculus_cd[\"start\"] = mus_musculus_cd[\"start\"].map(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mus_musculus_aca[\"start\"] = [int(aca_mm9_2_mm10.ix[i, 1]) if i in aca_mm9_2_mm10.index\n",
    "                            else np.nan for i in mus_musculus_aca.snor_id]\n",
    "mus_musculus_aca[\"end\"] = [int(aca_mm9_2_mm10.ix[i, 2]) if i in aca_mm9_2_mm10.index\n",
    "                            else np.nan for i in mus_musculus_aca.snor_id]\n",
    "mus_musculus_aca = mus_musculus_aca.dropna(subset=[\"start\"])\n",
    "mus_musculus_aca[\"start\"] = mus_musculus_aca[\"start\"].map(int)\n",
    "mus_musculus_aca[\"end\"] = mus_musculus_aca[\"end\"].map(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mus_musculus_cd.to_csv(\"mouse/mmu_cd_box_snoRNAs_mm10.tab\",\n",
    "                       index=False,\n",
    "                       header=False,\n",
    "                       sep=\"\\t\",\n",
    "                       na_rep=\"NaN\")\n",
    "mus_musculus_aca.to_csv(\"mouse/mmu_haca_box_snoRNAs_mm10.tab\",\n",
    "                       index=False,\n",
    "                       header=False,\n",
    "                       sep=\"\\t\",\n",
    "                       na_rep=\"NaN\",\n",
    "                        encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Comments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###First bugfix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some of the xml files required adjustment for the sequence and coordinates. The database is not complete. In the end I decided to loosen the requirement for the snoRNAs that snoRNA sequence length has to match snoRNA coordinates length. This solution was faster and as I need the coordinates only for expression estimation is valid. Moreover, some coordinates were provided in hg19 and some with hg18.\n",
    " - SNORD11B\n",
    " It did not have any coordinates. I took them from hadi and exchanged them.\n",
    "I exchanged coordinates (adjusted them to the sequence) and added chrom\n",
    " - SNORD19B\n",
    " Coordinates where wrong (totally different than in UCSC)\n",
    " - SNORD105B\n",
    " There was no coordinates provided\n",
    " - SNORD111B\n",
    " Coordinates were wrong\n",
    " - SNORD12B\n",
    " There was no coordinates\n",
    " - SNORD127\n",
    " There was no coordinates\n",
    " - SNORD123\n",
    " There is no such snoRNA in snOPY\n",
    " - SNORD1B\n",
    " Coordinates were wrong\n",
    " - SNORD14E\n",
    " no such a snoRNA in snOPY\n",
    " - SNORD32B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD1C\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD4B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD36B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD11\n",
    " no such a gene (only snord11b present)\n",
    " - SNORD35B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD56B\n",
    " no such snoRNA\n",
    " - SNORD96B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD10\n",
    " no genomic coordinates provided\n",
    " - SNORD42B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD109B\n",
    " no coordinates provided\n",
    " - SNORD62B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD54\n",
    " no coordinates provided\n",
    " - SNORD12C\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD58B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD45B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD45C\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD38B (Homo_sapiens300664.xml)\n",
    " there was 3 different genes for this snoRNAs in snOPY. One of them was consistent with Hadi's\n",
    " coordinates (however in hg18 so I changed to hg19)\n",
    " - SNORD15B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD14A (Homo_sapiens300737.xml)\n",
    " There were two genes. As in the case of 38B I chosen these one with the consistent chromosome\n",
    " to Hadi's result\n",
    " - SNORD14C\n",
    " no such snoRNA\n",
    " - SNORD14D\n",
    " no such snoRNA\n",
    " - SNORD13\n",
    " no such snoRNA\n",
    " - snord50B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD18B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD18C (Homo_sapiens300521.xml)\n",
    " Three options were obtained. Chosen as previously.\n",
    " - SNORD5\n",
    " This one had even chromosome wrongly assigned.\n",
    " - SNORD83B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD91B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD49B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD103B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD59B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD121B\n",
    " no such snoRNA\n",
    " - SNORD88C\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD88B\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD121A\n",
    " no such snoRNA\n",
    " - SNORD124\n",
    " no such snoRNA\n",
    " - SNORD125\n",
    " no such snoRNA\n",
    " - SNORD58C\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD126\n",
    " no such snoRNA\n",
    " - SNORD108\n",
    " wrong coordinates (from hg18)\n",
    " - SNORD97\n",
    " no such snoRNA\n",
    "\n",
    "\n",
    "Those are only snoRNAs that do not have \"-\" in the name ie. without known subform.\n",
    "\n",
    "Additionally following snoRNAs did not have coordinates:\n",
    " - Homo_sapiens300722    SNORA11B_sapiens300722\n",
    " - Homo_sapiens300762     SNORD11_sapiens300762\n",
    " - Homo_sapiens300764     SNORA34_sapiens300764\n",
    " - Homo_sapiens300765     SNORA78_sapiens300765\n",
    " - Homo_sapiens300767    SNORA36B_sapiens300767\n",
    " - Homo_sapiens300775    SCARNA14_sapiens300775\n",
    "\n",
    "and they were filled by hand."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Second bugfix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is disagreement between some of the targets found and database:\n",
    " - position 1805 (in snOPY) for RNA18S modified by SNORD20:\n",
    "we find the modification on the position 1806:\n",
    "RNA18S\t1792\t1832\t1625944_1:SNORD20_sapiens300509:8\t34\t+\tSNORD20_sapiens300509\tAGACGGTCGAACTTGACTATCTAGAGGAAGTAAAAGTCGT\t-32.0\t((((((((((((((((((((.&))))))))))))))))))))\tAACTTmGACTATCTAGAGGAAG&TTCCTCTAGATAGTCAAGTT\tD\t1806\t0.0\t47454.617287\n",
    "\n",
    "According to snoOPY this is 1805: GAACTmTGACT but according to snoRNABase and [1] this is as in our prediction.\n",
    "\n",
    "\n",
    " - The same situation occurs in the position 1703 (in snOPY) in RNA18S modified by SNORD43. According to our data the modified position\n",
    " is 1705:\n",
    "RNA18S\t1682\t1722\t11405452_1:SNORD43_sapiens300686:N\t30\t+\tSNORD43_sapiens300686\tTCCCTGCCCTTTGTACACACCGCCCGTCGCTACTACCGAT\t-21.9\t.(((((((((.&.))))))))).\tACCGCmCCGTCG&TGACGGGCGGA\tD'\t1705\t0.0\t20878.6803211\n",
    "\n",
    " What is confirmed in [1] and in snoRNABase.\n",
    "\n",
    "\n",
    " - The same for position 1678 modified by SNORD82 in RNA18S in snOPY (according to to snOPY modified nuc is T in sequence but A in snoRNA entry!!!).\n",
    " Our data says that it is position 1680 and the nucleotide is A:\n",
    "RNA18S\t1657\t1697\t2868101_1:SNORD82_sapiens300510:26\t52\t+\tSNORD82_sapiens300510\tGGGTCATAAGCTTGCGTTGATTAAGTCCCTGCCCTTTGTA\t-16.6\t.(((((((((((.&.))))))))))).\tGATTAmAGTCCCTG&AAGGGACTTAATA\tD'\t1680\t0.0\t72860.9418862\n",
    "\n",
    "This position (according to the sequence) is confirmed in snoRNABase.\n",
    "\n",
    " - CD boxes of SNORD25 are wrongly annotated: they should be 59 and 34 for D and D' box respectively, instead of 60 and 35. This mistakes causes that predicted \n",
    " nucleotide modification is 1489 and the true modification site is located 1490 (of course the modified nucleotide is also wrong).\n",
    "\n",
    "\n",
    "Conclusions:\n",
    " - There is a lot of mistakes in the snOPY database. There is bad modification position annotation and mixed genomes used."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Bibliography"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Kiss-László, Z., Henry, Y., Bachellerie, J. P., Caizergues-Ferrer, M. & Kiss, T. Site-specific ribose methylation of preribosomal RNA: a novel function for small nucleolar RNAs. Cell 85, 1077–1088 (1996)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "/*\n",
       " * example.css\n",
       " * Copyright (C) 2014 Rafal Gumienny\n",
       " *\n",
       " * Distributed under terms of the GPL license.\n",
       " */\n",
       "\n",
       "<link href='http://fonts.googleapis.com/css?family=Alegreya+Sans:100,300,400,500,700,800,900,100italic,300italic,400italic,500italic,700italic,800italic,900italic' rel='stylesheet' type='text/css'>\n",
       "<link href='http://fonts.googleapis.com/css?family=Arvo:400,700,400italic' rel='stylesheet' type='text/css'>\n",
       "<link href='http://fonts.googleapis.com/css?family=PT+Mono' rel='stylesheet' type='text/css'>\n",
       "<link href='http://fonts.googleapis.com/css?family=Shadows+Into+Light' rel='stylesheet' type='text/css'>\n",
       "<link href='http://fonts.googleapis.com/css?family=Philosopher:400,700,400italic,700italic' rel='stylesheet' type='text/css'>\n",
       "\n",
       "<style>\n",
       "\n",
       "@font-face {\n",
       "\tfont-family: \"Computer Modern\";\n",
       "\tsrc: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');\n",
       "}\n",
       "\n",
       "/* div.cell { */\n",
       "\t/* width: 800px; */\n",
       "\t/* margin-left: auto; */\n",
       "\t/* margin-right: auto; */\n",
       "/* } */\n",
       "\n",
       "#notebook li {\n",
       "margin-top:0.6em;\n",
       "}\n",
       "\n",
       "\n",
       "/* [> draw border around running cells <] */\n",
       "div.cell.border-box-sizing.code_cell.running { \n",
       "\tborder: 3px solid #111;\n",
       "\tborder-color: #B22222;\n",
       "\tborder-radius: 5px;\n",
       "}\n",
       "\n",
       "div.text_cell_render{\n",
       "    font-family: Computer Modern, \"Helvetica Neue\", Arial, Helvetica, Geneva, sans-serif;\n",
       "    line-height: 145%;\n",
       "    font-size: 130%;\n",
       "    /* margin-left:10px; */\n",
       "    /* margin-right:10px; */\n",
       "}\n",
       "\n",
       "/* Formatting for header cells */\n",
       ".text_cell_render h1 {\n",
       "\tfont-family: 'Serif', sans-serif;\n",
       "\tfont-weight: 400;\n",
       "\tfont-size: 30pt;\n",
       "\tline-height: 100%;\n",
       "\t/* color: rgb(12,85,97); */\n",
       "\tcolor: #b22222;\n",
       "\tmargin-bottom: 0.1em;\n",
       "\tmargin-top: 0.1em;\n",
       "\tdisplay: block;\n",
       "}\t\n",
       ".text_cell_render h2 {\n",
       "\tfont-family: 'Serif', serif;\n",
       "\tfont-weight: 700;\n",
       "\tfont-size: 20pt;\n",
       "\tline-height: 100%;\n",
       "\tcolor: #00008B;\n",
       "\t/* color: rgb(12,85,97); */\n",
       "\t/* color: rgb(171,165,131); */\n",
       "\tmargin-bottom: 0.1em;\n",
       "\tmargin-top: 0.3em;\n",
       "\tdisplay: block;\n",
       "}\t\n",
       "\n",
       ".text_cell_render h3 {\n",
       "\tfont-family: 'Serif', serif;\n",
       "\tmargin-bottom: 0.1em;\n",
       "\tmargin-top: 0.3em;\n",
       "\t/* margin-top:12px; */\n",
       "\t/* margin-bottom: 1px; */\n",
       "\tfont-size: 15pt;\n",
       "\tfont-style: italic;\n",
       "\tcolor: #004b00;\n",
       "\t/* color: rgb(95,92,72); */\n",
       "}\n",
       "\n",
       ".text_cell_render h4 {\n",
       "\tfont-family: serif;\n",
       "\tmargin-bottom: 0.1em;\n",
       "\tmargin-top: 0.3em;\n",
       "\t/* margin-top:12px; */\n",
       "\t/* margin-bottom: 1px; */\n",
       "}\n",
       "\n",
       ".text_cell_render h5 {\n",
       "\tfont-family: sans-serif;\n",
       "\tfont-weight: 300;\n",
       "\tfont-size: 13pt;\n",
       "\tcolor: grey;\n",
       "\tfont-style: italic;\n",
       "\tmargin-bottom: .1em;\n",
       "\tmargin-top: 0.3em;\n",
       "\tdisplay: block;\n",
       "}\n",
       "\n",
       ".text_cell_render h6 {\n",
       "\tfont-family: sans-serif;\n",
       "\tfont-weight: 300;\n",
       "\tfont-size: 10pt;\n",
       "\tcolor: grey;\n",
       "\tmargin-bottom: 1px;\n",
       "\tmargin-top: 1px;\n",
       "}\n",
       "\n",
       ".CodeMirror{\n",
       "        font-family: 'DejaVu Sans Mono';\n",
       "        font-size: 100%;\n",
       "}\n",
       "\n",
       "table\n",
       "{\n",
       "    font-family: \"Lucida Sans Unicode\", \"Lucida Grande\", Sans-Serif;\n",
       "    font-size: 14px;\n",
       "    margin: 45px;\n",
       "    text-align: center;\n",
       "    border-collapse: collapse;\n",
       "}\n",
       "table th\n",
       "{\n",
       "    font-size: 15px;\n",
       "    font-weight: normal;\n",
       "    padding: 8px;\n",
       "    background: #DCDCDC;\n",
       "    border-top: 4px solid #aabcfe;\n",
       "    border-bottom: 1px solid #fff;\n",
       "    color: #000000;\n",
       "\ttext-align: center;\n",
       "}\n",
       "table td\n",
       "{\n",
       "    padding: 8px;\n",
       "    background: #FFFACD;\n",
       "    border-bottom: 1px solid #fff;\n",
       "    color: #000000;\n",
       "    border-top: 1px solid transparent;\n",
       "}\n",
       "table tr:hover td\n",
       "{\n",
       "    background: #d0dafd;\n",
       "    color: #000000;\n",
       "}\n",
       "\n",
       "</style>\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import re\n",
    "import glob\n",
    "import swalign\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pylab as pl\n",
    "import seaborn as sns\n",
    "from random import choice\n",
    "import xml.etree.ElementTree as ET\n",
    "from collections import OrderedDict\n",
    "from Bio import SeqIO\n",
    "from IPython.display import Image\n",
    "pl.rcParams['figure.figsize'] = (14, 10)\n",
    "pl.rcParams['ytick.labelsize'] = 20\n",
    "pl.rcParams['xtick.labelsize'] = 20\n",
    "pl.rcParams['axes.labelsize'] = 23\n",
    "pl.rcParams['legend.fontsize'] = 20\n",
    "sns.set_style('ticks')\n",
    "c1, c2, c3 = sns.color_palette(\"Set1\", 3)\n",
    "from IPython.core.display import HTML, Image\n",
    "css_file = '/import/bc2/home/zavolan/gumiennr/.ipython/profile_default/static/custom/custom.css'\n",
    "HTML(open(css_file, \"r\").read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def apped_assembly(df):\n",
    "    #\n",
    "    # prepare swalign parameters\n",
    "    #\n",
    "    scoring = swalign.NucleotideScoringMatrix(match=2,\n",
    "                                              mismatch=-5)\n",
    "    sw = swalign.LocalAlignment(scoring,\n",
    "                                gap_penalty=-6,\n",
    "                                gap_extension_penalty=-4)\n",
    "    hg19_score = []\n",
    "    hg18_score = []\n",
    "    assembly = []\n",
    "    for idx, row in df.iterrows():\n",
    "        try:\n",
    "            h19 = sw.align(row[6], row[7]).score\n",
    "        except AttributeError:\n",
    "            h19 = 0.0\n",
    "        try:\n",
    "            h18 = sw.align(row[6], row[8]).score\n",
    "        except AttributeError:\n",
    "            h18 = 0.0\n",
    "        hg19_score.append(h19)\n",
    "        hg18_score.append(h18)\n",
    "\n",
    "        if h19 > h18:\n",
    "            assembly.append(\"hg19\")\n",
    "        elif h19 < h18:\n",
    "            assembly.append(\"hg18\")\n",
    "        else:\n",
    "            assembly.append(\"ambiguous\")\n",
    "    df[\"hg19score\"] = hg19_score\n",
    "    df[\"hg18score\"] = hg18_score\n",
    "    df[\"assembly\"] = assembly\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def apped_mouse_assembly(df):\n",
    "    #\n",
    "    # prepare swalign parameters\n",
    "    #\n",
    "    scoring = swalign.NucleotideScoringMatrix(match=2,\n",
    "                                              mismatch=-5)\n",
    "    sw = swalign.LocalAlignment(scoring,\n",
    "                                gap_penalty=-6,\n",
    "                                gap_extension_penalty=-4)\n",
    "    mm9_score = []\n",
    "    mm10_score = []\n",
    "    assembly = []\n",
    "    for idx, row in df.iterrows():\n",
    "        try:\n",
    "            mm10 = sw.align(row[6], row[7]).score\n",
    "        except AttributeError:\n",
    "            mm10 = 0.0\n",
    "        try:\n",
    "            mm9 = sw.align(row[6], row[8]).score\n",
    "        except AttributeError:\n",
    "            mm9 = 0.0\n",
    "        mm10_score.append(mm10)\n",
    "        mm9_score.append(mm9)\n",
    "\n",
    "        if mm10 > mm9:\n",
    "            assembly.append(\"mm10\")\n",
    "        elif mm10 < mm9:\n",
    "            assembly.append(\"mm9\")\n",
    "        else:\n",
    "            assembly.append(\"ambiguous\")\n",
    "    df[\"mm10score\"] = mm10_score\n",
    "    df[\"mm9score\"] = mm9_score\n",
    "    df[\"assembly\"] = assembly\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
