

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Usage &mdash; snoRNAHybridSearch 1.0.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="snoRNAHybridSearch 1.0.0 documentation" href="index.html"/>
        <link rel="next" title="Pipeline flow" href="scripts.html"/>
        <link rel="prev" title="Instalation" href="installation.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> snoRNAHybridSearch
          

          
          </a>

          
            
            
              <div class="version">
                1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Instalation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#basic-usage">Basic usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="#preparing-input-files">Preparing input files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#preparing-config-file">Preparing config file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#example">Example</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="scripts.html">Pipeline flow</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">snoRNAHybridSearch</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Usage</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/usage.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<div class="section" id="basic-usage">
<h2>Basic usage<a class="headerlink" href="#basic-usage" title="Permalink to this headline">¶</a></h2>
<p>Command to launch the pipeline is as follows:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>python snoRNAHybridSearch.py run --config congig.ini --name-suffix name_of_the_run
</pre></div>
</div>
<p>All parameters for the script:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">snoRNAHybridSearch</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">{</span><span class="n">run</span><span class="p">,</span><span class="n">clean</span><span class="p">}</span> <span class="o">...</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Sub-commands:</dt>
<dd><dl class="first last docutils">
<dt><strong>run</strong></dt>
<dd><p class="first">Run a pipeline</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">snoRNAHybridSearch</span> <span class="n">run</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="o">--</span><span class="n">config</span> <span class="n">CONFIG</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">name</span><span class="o">-</span><span class="n">suffix</span> <span class="n">NAME_SUFFIX</span><span class="p">]</span>
                              <span class="p">[</span><span class="o">--</span><span class="nb">filter</span><span class="o">-</span><span class="n">multimappers</span><span class="p">]</span>
</pre></div>
</div>
<dl class="last docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--config</span></kbd></td>
<td>Config file</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--name-suffix=test_run</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Suffix to add to pipeline name in order to easily differentiate between different run, defaults to test_run</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--filter-multimappers=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Filter reads that map to multiple genomic locus with exception of reads that map also to canonical targets</td></tr>
</tbody>
</table>
</dd>
</dl>
</dd>
<dt><strong>clean</strong></dt>
<dd><p class="first">Clean after previous run</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">snoRNAHybridSearch</span> <span class="n">clean</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">y</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">backup</span><span class="p">]</span>
</pre></div>
</div>
<dl class="last docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-y=False</span>, <span class="option">--yes=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Force deletion of files.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--make-backup=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Instead of deleting file with results make its backup</td></tr>
</tbody>
</table>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
</div>
<div class="section" id="preparing-input-files">
<h2>Preparing input files<a class="headerlink" href="#preparing-input-files" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="preparing-config-file">
<h2>Preparing config file<a class="headerlink" href="#preparing-config-file" title="Permalink to this headline">¶</a></h2>
<p>Copy config_example.ini from MIRZA-G directory to your working directory (directory
where you want to perform calculation, WD):</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="nb">cd</span> Your/Working/Direcory
cp Path/To/snoRNAHybridSearch/config_example.ini config.ini
</pre></div>
</div>
<dl class="docutils">
<dt>Set all the necessary paths in your config.ini file as indicated in the comments inside the file. The most importand are:</dt>
<dd><ul class="first last simple">
<li><strong>motifs</strong>: &#8220;Path/To/miRNAs.fa&#8221; - abs path to an input fasta file with mi/siRNA sequences of length 21 or more</li>
<li><strong>seqs</strong>: &#8220;Path/To/MIRZA-G/data/UTR_Sequences.fa&#8221; - abs path to a fasta file with the UTR sequences from which the coordinate file will be generated (you can use 3&#8217;UTR sequences in the
MIRZA-G/data directory, for this file there are also alignments for conservation precalculated)</li>
<li><strong>mirza_binary</strong>: &#8220;MIRZA&#8221; - path to MIRZA binary (or how you invoke it in the bash)</li>
<li><strong>contrafold_binary</strong>: &#8220;contrafold&#8221; - path to CONTRAfold binary (or how you invoke in the bash)</li>
</ul>
</dd>
<dt>Models paths:</dt>
<dd><ul class="first last simple">
<li><strong>model_with_bls</strong>: &#8220;Path/To/MIRZA-G/data/glm-with-bls.bin&#8221; - abs path to the model with BLS (you can find it in the pipeline/data directory)</li>
<li><strong>model_without_bls</strong>: &#8220;Path/To/MIRZA-G/data/glm-without-bls.bin&#8221; - same as before</li>
</ul>
</dd>
<dt>Additionally when you would like to calculate with evolutionary conservation you have to make sure that the variable run_only_MIRZA in CalculateMIRZA task is set to “no” instead of “yes” and that you provide proper paths with aligned UTRs and evolutionary tree:</dt>
<dd><ul class="first last">
<li><p class="first"><strong>phylogenetic_tree</strong>: &#8220;Path/To/MIRZA-G/data/human_tree.nh&#8221; - abspath to provided phylogenetic tree</p>
</li>
<li><dl class="first docutils">
<dt><strong>alignment_directory</strong>: &#8220;Path/To/MIRZA-G/data/HumanAlignments/&#8221; - abspath to provided human alignments directory. If you downloaded package from CLIPz website</dt>
<dd><p class="first last">this directory is already in the MIRZA-G directory. If you downloaded from GitHub you have to download it additionally.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<p>If you would like to run it on cluster follow instructions in the configuration file and ask your admin what parameters you need to set
up before (like DRMAA path, modules necessary, queues names etc.). All these parameters can be set up in config.ini.</p>
<p>To run it locally it takes ~70 to 90 seconds for one miRNA without conservation calculation and ~170 seconds with calculation (This
might be substantial amount of time (up to half an hour per miRNA) for worse processors).</p>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>To test the pipeline go to the tests directory and run:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="nb">cd</span> Path/To/snoRNAHybridSearch/tests
bash rg_run_test.sh <span class="nb">help</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Usage: rg_run_test.sh clean/run [&#8216;CONTRAfold/binary/path&#8217;]</p>
</div>
<p>And if you have installed MIRZA and CONTRAfold to default locations (MIRZA and contrafold) run:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>bash rg_run_test.sh run
</pre></div>
</div>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="scripts.html" class="btn btn-neutral float-right" title="Pipeline flow" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral" title="Instalation" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Rafal Gumienny.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>