

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Usage &mdash; snoRNAHybridSearch 1.0.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="snoRNAHybridSearch 1.0.0 documentation" href="index.html"/>
        <link rel="next" title="Pipeline flow" href="scripts.html"/>
        <link rel="prev" title="Instalation" href="installation.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> snoRNAHybridSearch
          

          
          </a>

          
            
            
              <div class="version">
                1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Instalation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#basic-usage">Basic usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="#preparing-input-files">Preparing input files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#preparing-config-file">Preparing config file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#example">Example</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="scripts.html">Pipeline flow</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">snoRNAHybridSearch</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Usage</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/usage.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<div class="section" id="basic-usage">
<h2>Basic usage<a class="headerlink" href="#basic-usage" title="Permalink to this headline">¶</a></h2>
<p>Command to launch the pipeline is as follows:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>python snoRNAHybridSearch.py run --config congig.ini --name-suffix name_of_the_run
</pre></div>
</div>
<p>All parameters for the script:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">snoRNAHybridSearch</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">{</span><span class="n">run</span><span class="p">,</span><span class="n">clean</span><span class="p">}</span> <span class="o">...</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Sub-commands:</dt>
<dd><dl class="first last docutils">
<dt><strong>run</strong></dt>
<dd><p class="first">Run a pipeline</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">snoRNAHybridSearch</span> <span class="n">run</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="o">--</span><span class="n">config</span> <span class="n">CONFIG</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">name</span><span class="o">-</span><span class="n">suffix</span> <span class="n">NAME_SUFFIX</span><span class="p">]</span>
                              <span class="p">[</span><span class="o">--</span><span class="nb">filter</span><span class="o">-</span><span class="n">multimappers</span><span class="p">]</span>
</pre></div>
</div>
<dl class="last docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--config</span></kbd></td>
<td>Config file</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--name-suffix=test_run</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Suffix to add to pipeline name in order to easily differentiate between different run, defaults to test_run</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--filter-multimappers=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Filter reads that map to multiple genomic locus with exception of reads that map also to canonical targets</td></tr>
</tbody>
</table>
</dd>
</dl>
</dd>
<dt><strong>clean</strong></dt>
<dd><p class="first">Clean after previous run</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">snoRNAHybridSearch</span> <span class="n">clean</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">y</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">backup</span><span class="p">]</span>
</pre></div>
</div>
<dl class="last docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-y=False</span>, <span class="option">--yes=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Force deletion of files.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--make-backup=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Instead of deleting file with results make its backup</td></tr>
</tbody>
</table>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
</div>
<div class="section" id="preparing-input-files">
<h2>Preparing input files<a class="headerlink" href="#preparing-input-files" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="preparing-config-file">
<h2>Preparing config file<a class="headerlink" href="#preparing-config-file" title="Permalink to this headline">¶</a></h2>
<p>Copy config_example.ini from MIRZA-G directory to your working directory (directory
where you want to perform calculation, WD):</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="nb">cd</span> Your/Working/Direcory
cp Path/To/snoRNAHybridSearch/config_example.ini config.ini
</pre></div>
</div>
<dl class="docutils">
<dt>Set all the necessary paths in your config.ini file as indicated in the comments inside the file. The most importand are:</dt>
<dd><ul class="first last simple">
<li><strong>unmapped_reads</strong>: &#8220;Path/To/unmapped_reads.fa&#8221; - an abs path to an input FASTA file with sequences that were unmapped in sequencing experiment</li>
<li><strong>bed_for_index</strong>: &#8220;Path/To/mapped_reads.bed&#8221; - abs path to a BED file with the positions of mapped reads in the experiment</li>
<li><strong>PLEXY_bin</strong>: &#8220;Path/To/plexy.pl&#8221; - path to PLEXY binary (or how you invoke it in the bash)</li>
<li><strong>contrafold_binary</strong>: &#8220;contrafold&#8221; - path to CONTRAfold binary (or how you invoke in the bash)</li>
</ul>
</dd>
<dt>Model path:</dt>
<dd><ul class="first last simple">
<li><strong>model</strong>: &#8220;Path/To/snoRNAHybridSearch/data/model.bin&#8221; - abs path to the model used to calculate probability (you can find it in the pipeline/data directory)</li>
</ul>
</dd>
<dt>snoRNA table:</dt>
<dd><ul class="first last simple">
<li><strong>snoRNAs</strong>: &#8220;Absolute/Path/To/snoRNAs_table.tab&#8221; - abs path to the table containing all the necessary information abut snoRNAs. This table is provided with
pipeline and for human it is located in the data/human/snoRNAs/snoRNAs_table.tab. We have also prepared the table for mouse located in the data/mouse/snoRNAs/snoRNAs_table.tab.
You can also prepare your own snoRNA input - please follow the conventions in the table and pay attention to columns described in the data/Columns_description file.</li>
</ul>
</dd>
<dt>Additional &#8220;chromosomes&#8221;:</dt>
<dd><ul class="first last simple">
<li>This files has to be also split into separate FASTA sequences and those sequences has
to be put into directory with genome. By default, genome directory that can be downloaded additionally contains these sequences already prepared.</li>
<li><strong>rRNAs</strong>: &#8220;Absolute/Path/To/rRNAs.fa&#8221; # rRNA sequences. This is provided with the pipeline in data directory, although own can be used. The location for human is
data/human/target_RNAs/rRNAs_leipzig_hsa.fa and for mouse data/mouse/rRNAs_mmu.fa.</li>
<li><strong>tRNAs</strong>: &#8220;Absolute/Path/To/tRNAs.fa&#8221; # tRNA sequences. This is provided with the pipeline in data directory, although own can be used. The location for human is
data/human/target_RNAs/tRNAs_raw.fa and for mouse data/mouse/tRNAs_mmu.fa</li>
<li><strong>snRNAs</strong>: &#8220;Absolute/Path/To/snRNAs.fa&#8221; # snRNA sequences. This is provided with the pipeline in data directory, although own can be used. The location for human is
data/human/target_RNAs/snRNAs_leipzig_hsa.fa and for mouse data/mouse/snRNAs_mmu.fa</li>
</ul>
</dd>
<dt>Annotation files:</dt>
<dd><ul class="first last">
<li><p class="first">Annotation files are used to annotate found target positions. They are generated from corresponding ENSEMBL/GENECODE
gff3 files or downloaded from NCBI. These files can be found in the annotations subdirectory in given species data directory.</p>
</li>
<li><p class="first"><strong>annotations_genes</strong>: &#8220;Absolute/Path/To/Annotations/genes.gff3&#8221;. This file is generated from ENSEMBL/GENECODE file and contains information
abut genes - not transcripts:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="mi">1</span>   <span class="n">pseudogene</span>  <span class="n">gene</span>    <span class="mi">11869</span>   <span class="mi">14412</span>   <span class="o">.</span>   <span class="o">+</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;ENSG00000223972&quot;</span><span class="p">;</span> <span class="n">gene_name</span> <span class="s2">&quot;DDX11L1&quot;</span><span class="p">;</span> <span class="n">gene_source</span> <span class="s2">&quot;ensembl_havana&quot;</span><span class="p">;</span> <span class="n">gene_biotype</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">;</span>
<span class="mi">1</span>   <span class="n">pseudogene</span>  <span class="n">gene</span>    <span class="mi">14363</span>   <span class="mi">29806</span>   <span class="o">.</span>   <span class="o">-</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;ENSG00000227232&quot;</span><span class="p">;</span> <span class="n">gene_name</span> <span class="s2">&quot;WASH7P&quot;</span><span class="p">;</span> <span class="n">gene_source</span> <span class="s2">&quot;ensembl_havana&quot;</span><span class="p">;</span> <span class="n">gene_biotype</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">;</span>
<span class="mi">1</span>   <span class="n">lincRNA</span> <span class="n">gene</span>    <span class="mi">29554</span>   <span class="mi">31109</span>   <span class="o">.</span>   <span class="o">+</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;ENSG00000243485&quot;</span><span class="p">;</span> <span class="n">gene_name</span> <span class="s2">&quot;MIR1302-10&quot;</span><span class="p">;</span> <span class="n">gene_source</span> <span class="s2">&quot;ensembl_havana&quot;</span><span class="p">;</span> <span class="n">gene_biotype</span> <span class="s2">&quot;lincRNA&quot;</span><span class="p">;</span>
<span class="mi">1</span>   <span class="n">lincRNA</span> <span class="n">gene</span>    <span class="mi">34554</span>   <span class="mi">36081</span>   <span class="o">.</span>   <span class="o">-</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;ENSG00000237613&quot;</span><span class="p">;</span> <span class="n">gene_name</span> <span class="s2">&quot;FAM138A&quot;</span><span class="p">;</span> <span class="n">gene_source</span> <span class="s2">&quot;ensembl_havana&quot;</span><span class="p">;</span> <span class="n">gene_biotype</span> <span class="s2">&quot;lincRNA&quot;</span><span class="p">;</span>
<span class="mi">1</span>   <span class="n">pseudogene</span>  <span class="n">gene</span>    <span class="mi">52473</span>   <span class="mi">54936</span>   <span class="o">.</span>   <span class="o">+</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;ENSG00000268020&quot;</span><span class="p">;</span> <span class="n">gene_name</span> <span class="s2">&quot;OR4G4P&quot;</span><span class="p">;</span> <span class="n">gene_source</span> <span class="s2">&quot;ensembl_havana&quot;</span><span class="p">;</span> <span class="n">gene_biotype</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">;</span>
<span class="mi">1</span>   <span class="n">pseudogene</span>  <span class="n">gene</span>    <span class="mi">62948</span>   <span class="mi">63887</span>   <span class="o">.</span>   <span class="o">+</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;ENSG00000240361&quot;</span><span class="p">;</span> <span class="n">gene_name</span> <span class="s2">&quot;OR4G11P&quot;</span><span class="p">;</span> <span class="n">gene_source</span> <span class="s2">&quot;havana&quot;</span><span class="p">;</span> <span class="n">gene_biotype</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">;</span>
<span class="mi">1</span>   <span class="n">protein_coding</span>  <span class="n">gene</span>    <span class="mi">69091</span>   <span class="mi">70008</span>   <span class="o">.</span>   <span class="o">+</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;ENSG00000186092&quot;</span><span class="p">;</span> <span class="n">gene_name</span> <span class="s2">&quot;OR4F5&quot;</span><span class="p">;</span> <span class="n">gene_source</span> <span class="s2">&quot;ensembl_havana&quot;</span><span class="p">;</span> <span class="n">gene_biotype</span> <span class="s2">&quot;protein_coding&quot;</span><span class="p">;</span>
<span class="mi">1</span>   <span class="n">lincRNA</span> <span class="n">gene</span>    <span class="mi">89295</span>   <span class="mi">133566</span>  <span class="o">.</span>   <span class="o">-</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;ENSG00000238009&quot;</span><span class="p">;</span> <span class="n">gene_name</span> <span class="s2">&quot;RP11-34P13.7&quot;</span><span class="p">;</span> <span class="n">gene_source</span> <span class="s2">&quot;havana&quot;</span><span class="p">;</span> <span class="n">gene_biotype</span> <span class="s2">&quot;lincRNA&quot;</span><span class="p">;</span>
<span class="mi">1</span>   <span class="n">lincRNA</span> <span class="n">gene</span>    <span class="mi">89551</span>   <span class="mi">91105</span>   <span class="o">.</span>   <span class="o">-</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;ENSG00000239945&quot;</span><span class="p">;</span> <span class="n">gene_name</span> <span class="s2">&quot;RP11-34P13.8&quot;</span><span class="p">;</span> <span class="n">gene_source</span> <span class="s2">&quot;havana&quot;</span><span class="p">;</span> <span class="n">gene_biotype</span> <span class="s2">&quot;lincRNA&quot;</span><span class="p">;</span>
<span class="mi">1</span>   <span class="n">pseudogene</span>  <span class="n">gene</span>    <span class="mi">131025</span>  <span class="mi">134836</span>  <span class="o">.</span>   <span class="o">+</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;ENSG00000233750&quot;</span><span class="p">;</span> <span class="n">gene_name</span> <span class="s2">&quot;CICP27&quot;</span><span class="p">;</span> <span class="n">gene_source</span> <span class="s2">&quot;havana&quot;</span><span class="p">;</span> <span class="n">gene_biotype</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">;</span>
</pre></div>
</div>
</li>
<li><p class="first"><strong>annotations_regions</strong>: &#8220;Absolute/Path/To/Annotations/regions.gff3&#8221;. This file is generated from ENSEMBL/GENECODE file and contains information
abut the regions in the genes and transcripts like introns, exons, and UTRS:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="mi">1</span>   <span class="n">ensembl_havana</span>  <span class="n">exon</span>    <span class="mi">69091</span>   <span class="mi">70008</span>   <span class="o">.</span>   <span class="o">+</span>   <span class="o">.</span>   <span class="n">Parent</span><span class="o">=</span><span class="n">mRNA_ENST00000335137</span>
<span class="mi">1</span>   <span class="n">ensembl_havana</span>  <span class="n">CDS</span> <span class="mi">69091</span>   <span class="mi">70008</span>   <span class="o">.</span>   <span class="o">+</span>   <span class="o">.</span>   <span class="n">Parent</span><span class="o">=</span><span class="n">mRNA_ENST00000335137</span>
<span class="mi">1</span>   <span class="n">ensembl</span> <span class="n">exon</span>    <span class="mi">134901</span>  <span class="mi">135802</span>  <span class="o">.</span>   <span class="o">-</span>   <span class="o">.</span>   <span class="n">Parent</span><span class="o">=</span><span class="n">mRNA_ENST00000423372</span>
<span class="mi">1</span>   <span class="n">ensembl</span> <span class="n">intron</span>  <span class="mi">135803</span>  <span class="mi">137620</span>  <span class="o">.</span>   <span class="o">-</span>   <span class="o">.</span>   <span class="n">Parent</span><span class="o">=</span><span class="n">mRNA_ENST00000423372</span>
<span class="mi">1</span>   <span class="n">ensembl</span> <span class="n">exon</span>    <span class="mi">137621</span>  <span class="mi">139379</span>  <span class="o">.</span>   <span class="o">-</span>   <span class="o">.</span>   <span class="n">Parent</span><span class="o">=</span><span class="n">mRNA_ENST00000423372</span>
<span class="mi">1</span>   <span class="n">ensembl</span> <span class="n">three_prime_UTR</span> <span class="mi">134901</span>  <span class="mi">135802</span>  <span class="o">.</span>   <span class="o">-</span>   <span class="o">.</span>   <span class="n">Parent</span><span class="o">=</span><span class="n">mRNA_ENST00000423372</span>
<span class="mi">1</span>   <span class="n">ensembl</span> <span class="n">three_prime_UTR</span> <span class="mi">137621</span>  <span class="mi">138529</span>  <span class="o">.</span>   <span class="o">-</span>   <span class="o">.</span>   <span class="n">Parent</span><span class="o">=</span><span class="n">mRNA_ENST00000423372</span>
<span class="mi">1</span>   <span class="n">ensembl</span> <span class="n">CDS</span> <span class="mi">138530</span>  <span class="mi">139309</span>  <span class="o">.</span>   <span class="o">-</span>   <span class="o">.</span>   <span class="n">Parent</span><span class="o">=</span><span class="n">mRNA_ENST00000423372</span>
<span class="mi">1</span>   <span class="n">ensembl</span> <span class="n">five_prime_UTR</span>  <span class="mi">139310</span>  <span class="mi">139379</span>  <span class="o">.</span>   <span class="o">-</span>   <span class="o">.</span>   <span class="n">Parent</span><span class="o">=</span><span class="n">mRNA_ENST00000423372</span>
<span class="mi">1</span>   <span class="n">ensembl_havana</span>  <span class="n">exon</span>    <span class="mi">367640</span>  <span class="mi">368634</span>  <span class="o">.</span>   <span class="o">+</span>   <span class="o">.</span>   <span class="n">Parent</span><span class="o">=</span><span class="n">mRNA_ENST00000426406</span>
</pre></div>
</div>
</li>
<li><p class="first"><strong>annotations_repeats</strong>: &#8220;Absolute/Path/To/Annotations/repeats.gtf&#8221;. It is a file downloaded from NCBI table browser:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">chr1</span>    <span class="n">hg19_rmsk</span>   <span class="n">exon</span>    <span class="mi">16777161</span>    <span class="mi">16777470</span>    <span class="mf">2147.000000</span> <span class="o">+</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;AluSp&quot;</span><span class="p">;</span> <span class="n">transcript_id</span> <span class="s2">&quot;AluSp&quot;</span><span class="p">;</span>
<span class="n">chr1</span>    <span class="n">hg19_rmsk</span>   <span class="n">exon</span>    <span class="mi">25165801</span>    <span class="mi">25166089</span>    <span class="mf">2626.000000</span> <span class="o">-</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;AluY&quot;</span><span class="p">;</span> <span class="n">transcript_id</span> <span class="s2">&quot;AluY&quot;</span><span class="p">;</span>
<span class="n">chr1</span>    <span class="n">hg19_rmsk</span>   <span class="n">exon</span>    <span class="mi">33553607</span>    <span class="mi">33554646</span>    <span class="mf">626.000000</span>  <span class="o">+</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;L2b&quot;</span><span class="p">;</span> <span class="n">transcript_id</span> <span class="s2">&quot;L2b&quot;</span><span class="p">;</span>
<span class="n">chr1</span>    <span class="n">hg19_rmsk</span>   <span class="n">exon</span>    <span class="mi">50330064</span>    <span class="mi">50332153</span>    <span class="mf">12545.000000</span>    <span class="o">+</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;L1PA10&quot;</span><span class="p">;</span> <span class="n">transcript_id</span> <span class="s2">&quot;L1PA10&quot;</span><span class="p">;</span>
<span class="n">chr1</span>    <span class="n">hg19_rmsk</span>   <span class="n">exon</span>    <span class="mi">58720068</span>    <span class="mi">58720973</span>    <span class="mf">8050.000000</span> <span class="o">-</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;L1PA2&quot;</span><span class="p">;</span> <span class="n">transcript_id</span> <span class="s2">&quot;L1PA2&quot;</span><span class="p">;</span>
<span class="n">chr1</span>    <span class="n">hg19_rmsk</span>   <span class="n">exon</span>    <span class="mi">75496181</span>    <span class="mi">75498100</span>    <span class="mf">10586.000000</span>    <span class="o">+</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;L1MB7&quot;</span><span class="p">;</span> <span class="n">transcript_id</span> <span class="s2">&quot;L1MB7&quot;</span><span class="p">;</span>
<span class="n">chr1</span>    <span class="n">hg19_rmsk</span>   <span class="n">exon</span>    <span class="mi">83886031</span>    <span class="mi">83886750</span>    <span class="mf">980.000000</span>  <span class="o">-</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;ERVL-E-int&quot;</span><span class="p">;</span> <span class="n">transcript_id</span> <span class="s2">&quot;ERVL-E-int&quot;</span><span class="p">;</span>
<span class="n">chr1</span>    <span class="n">hg19_rmsk</span>   <span class="n">exon</span>    <span class="mi">100662896</span>   <span class="mi">100663391</span>   <span class="mf">1422.000000</span> <span class="o">-</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;L2a&quot;</span><span class="p">;</span> <span class="n">transcript_id</span> <span class="s2">&quot;L2a&quot;</span><span class="p">;</span>
<span class="n">chr1</span>    <span class="n">hg19_rmsk</span>   <span class="n">exon</span>    <span class="mi">117440427</span>   <span class="mi">117440514</span>   <span class="mf">532.000000</span>  <span class="o">+</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;L1ME1&quot;</span><span class="p">;</span> <span class="n">transcript_id</span> <span class="s2">&quot;L1ME1&quot;</span><span class="p">;</span>
<span class="n">chr1</span>    <span class="n">hg19_rmsk</span>   <span class="n">exon</span>    <span class="mi">117440495</span>   <span class="mi">117441457</span>   <span class="mf">4025.000000</span> <span class="o">+</span>   <span class="o">.</span>   <span class="n">gene_id</span> <span class="s2">&quot;L1ME1&quot;</span><span class="p">;</span> <span class="n">transcript_id</span> <span class="s2">&quot;L1ME1_dup1&quot;</span><span class="p">;</span>
</pre></div>
</div>
</li>
</ul>
</dd>
<dt>Others:</dt>
<dd><ul class="first last simple">
<li><strong>reads_per_file</strong>: number of reads in the split files</li>
<li><strong>anchor_length</strong>: the lenght of the &#8220;seed&#8221; prepared from snoRNAs which will be searched initially in the unmapped sequences</li>
<li>If you would like to run it on cluster follow instructions in the configuration file and ask your admin what parameters you need to set
up before (like DRMAA path, modules necessary, queues names etc.). All these parameters can be set up in config.ini. To run it locally it
takes ~70 to 90 seconds for one miRNA without conservation calculation and ~170 seconds with calculation (This
might be substantial amount of time (up to half an hour per miRNA) for worse processors).</li>
</ul>
</dd>
</dl>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>To test the pipeline go to the tests directory and run:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="nb">cd</span> Path/To/snoRNAHybridSearch/tests
bash rg_run_test.sh <span class="nb">help</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Usage: rg_run_test.sh clean/run [&#8216;CONTRAfold/binary/path&#8217;]</p>
</div>
<p>And if you have installed MIRZA and CONTRAfold to default locations (MIRZA and contrafold) run:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>bash rg_run_test.sh run
</pre></div>
</div>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="scripts.html" class="btn btn-neutral float-right" title="Pipeline flow" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral" title="Instalation" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Rafal Gumienny.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>